// @jovotech/output/dist/esm2015/models/Message.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { IsSomeValid, isString } from '..';
export const IsValidMessageString = () => IsSomeValid({
    keys: ['speech', 'text'],
    validate: (value) => {
        if (!isString(value)) {
            return '$property must be a string';
        }
        if (!value) {
            return '$property should not be empty';
        }
        return;
    },
});
export class Message {
}
__decorate([
    IsValidMessageString(),
    __metadata("design:type", String)
], Message.prototype, "speech", void 0);
__decorate([
    IsValidMessageString(),
    __metadata("design:type", String)
], Message.prototype, "text", void 0);
export class SpeechMessage extends Message {
}
export class TextMessage extends Message {
}
