// @jovotech/platform-alexa/dist/esm2015/api/AlexaApi.js
import { axios, JovoError, } from '@jovotech/framework';
import { URL } from 'url';
export var AlexaApiErrorCode;
(function (AlexaApiErrorCode) {
    AlexaApiErrorCode["PARSE_ERROR"] = "PARSE_ERROR";
    AlexaApiErrorCode["ACCESS_NOT_REQUESTED"] = "ACCESS_NOT_REQUESTED";
    AlexaApiErrorCode["NO_USER_PERMISSION"] = "NO_USER_PERMISSION";
    AlexaApiErrorCode["NO_SKILL_PERMISSION"] = "NO_SKILL_PERMISSION";
    AlexaApiErrorCode["LIST_NOT_FOUND"] = "LIST_NOT_FOUND";
    AlexaApiErrorCode["ITEM_NOT_FOUND"] = "ITEM_NOT_FOUND";
    AlexaApiErrorCode["DEVICE_NOT_SUPPORTED"] = "DEVICE_NOT_SUPPORTED";
    AlexaApiErrorCode["ALERT_NOT_FOUND"] = "ALERT_NOT_FOUND";
    AlexaApiErrorCode["ERROR"] = "ERROR";
})(AlexaApiErrorCode || (AlexaApiErrorCode = {}));
export class AlexaApiError extends JovoError {
    constructor(options) {
        super(Object.assign(Object.assign({}, options), { message: `Request to Alexa API failed: ${options.message}` }));
        this.code = options.code;
    }
}
/**
 * Parses options and sends a request to the specified API endpoint
 * @param options - API options
 */
export async function sendApiRequest(options) {
    const url = new URL(options.path, options.endpoint);
    const config = {
        url: url.href,
        data: options.data,
        method: options.method || 'GET',
        headers: Object.assign({ 'Content-Type': 'application/json', 'Authorization': `Bearer ${options.permissionToken}` }, options.headers),
        params: options.params,
    };
    return await axios(config);
}
