// @jovotech/platform-alexa/dist/esm2015/api/ListApi.js
import { JovoError } from '@jovotech/framework';
import { AlexaApiError, AlexaApiErrorCode, sendApiRequest, } from './AlexaApi';
export async function getLists(apiEndpoint, permissionToken) {
    const options = {
        endpoint: apiEndpoint,
        path: `/v2/householdlists`,
        permissionToken,
        method: 'GET',
    };
    try {
        const response = await sendApiRequest(options);
        return response.data.lists;
    }
    catch (error) {
        handleListApiErrors(error);
    }
    throw new Error('Unexpected error.');
}
/**
 * Returns the type of the list
 * @param listId List to check
 * @returns the type of the list
 */
export function getTypeOfList(listId) {
    const decodedListId = Buffer.from(listId, 'base64').toString('utf8');
    if (decodedListId.endsWith('-SHOPPING_ITEM')) {
        return 'shopping-list';
    }
    else {
        return 'todo-list';
    }
}
export async function getListItems(listId, itemIds, apiEndpoint, permissionToken) {
    return Promise.all(itemIds.map((itemId) => getListItem(listId, itemId, apiEndpoint, permissionToken)));
}
export async function getListItem(listId, itemId, apiEndpoint, permissionToken) {
    const options = {
        endpoint: apiEndpoint,
        path: `/v2/householdlists/${listId}/items/${itemId}`,
        permissionToken,
        method: 'GET',
    };
    try {
        const response = await sendApiRequest(options);
        return response.data;
    }
    catch (error) {
        handleListApiErrors(error);
    }
    throw new Error('Unexpected error.');
}
/**
 * Meaning of error-codes can be found here: https://developer.amazon.com/en-US/docs/alexa/list-skills/list-management-api-reference.html#get-list-item-http-status-codes
 * @param error Error to handle
 */
function handleListApiErrors(error) {
    var _a, _b;
    if (error.isAxiosError) {
        const { message } = (_a = error.response) === null || _a === void 0 ? void 0 : _a.data;
        let errorCode = AlexaApiErrorCode.ERROR;
        const status = (_b = error.response) === null || _b === void 0 ? void 0 : _b.status;
        if (status === 401) {
            errorCode = AlexaApiErrorCode.NO_USER_PERMISSION;
        }
        if (status === 404) {
            errorCode = AlexaApiErrorCode.LIST_NOT_FOUND;
        }
        throw new AlexaApiError({ message, code: errorCode });
    }
    throw new JovoError({ message: error.message });
}
