// @jovotech/platform-alexa/dist/esm2015/api/PersonProfileApi.js
import { JovoError } from '@jovotech/framework';
import { AlexaApiError, AlexaApiErrorCode, sendApiRequest } from './AlexaApi';
export var PersonProfileProperty;
(function (PersonProfileProperty) {
    PersonProfileProperty["NAME"] = "name";
    PersonProfileProperty["GIVEN_NAME"] = "givenName";
    PersonProfileProperty["MOBILE_NUMBER"] = "mobileNumber";
})(PersonProfileProperty || (PersonProfileProperty = {}));
/**
 * Sends a request to Amazon's Person Profile API for getting profile information
 * @param profileProperty - The profile property which determines the final API endpoint url
 * @param apiEndpoint - API endpoint, differs on the geographic location of the skill
 * @param permissionToken - Token to authorize the request
 * @see {@link https://developer.amazon.com/en-US/docs/alexa/custom-skills/request-recognized-speaker-contact-information.html Request Recognized Speaker Contact Information}
 */
export async function sendPersonProfileApiRequest(profileProperty, apiEndpoint, permissionToken) {
    const options = {
        endpoint: apiEndpoint,
        path: `/v2/persons/~current/profile/${profileProperty}`,
        permissionToken,
    };
    try {
        const response = await sendApiRequest(options);
        return response.data;
    }
    catch (error) {
        if (error.isAxiosError) {
            const { message, code } = error.response.data;
            let errorCode = AlexaApiErrorCode.ERROR;
            // User needs to grant access in app
            if (message === 'The authentication token is not valid.' ||
                message === 'Access to this resource has not yet been requested.' ||
                (code === 'ACCESS_DENIED' && message === 'Access denied with reason: ACCESS_NOT_REQUESTED')) {
                errorCode = AlexaApiErrorCode.NO_USER_PERMISSION;
            }
            // Dev needs to set correct permissions in ASK console
            if (message === 'Access to this resource cannot be requested.' ||
                (code === 'ACCESS_DENIED' && message === 'Access denied with reason: FORBIDDEN')) {
                errorCode = AlexaApiErrorCode.NO_SKILL_PERMISSION;
            }
            throw new AlexaApiError({ message, code: errorCode });
        }
        throw new JovoError({ message: error.message });
    }
}
