// @jovotech/framework/dist/esm5/JovoDevice.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
export var Capability;
(function (Capability) {
    Capability["Screen"] = "SCREEN";
    Capability["Audio"] = "AUDIO";
    Capability["LongformAudio"] = "LONGFORM_AUDIO";
    Capability["Video"] = "VIDEO";
})(Capability || (Capability = {}));
var JovoDevice = /** @class */ (function () {
    function JovoDevice(jovo) {
        this.jovo = jovo;
        this.capabilities = jovo.$request.getDeviceCapabilities() || [];
    }
    JovoDevice.prototype.supports = function (capability) {
        return this.capabilities.includes(capability);
    };
    JovoDevice.prototype.toJSON = function () {
        return __assign(__assign({}, this), { jovo: undefined });
    };
    return JovoDevice;
}());
export { JovoDevice };
