// @jovotech/framework/dist/esm5/JovoUser.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var JovoUser = /** @class */ (function () {
    function JovoUser(jovo) {
        this.jovo = jovo;
        this.createdAt = new Date();
        this.updatedAt = new Date();
        this.data = {};
        this.isNew = true;
    }
    Object.defineProperty(JovoUser.prototype, "accessToken", {
        get: function () {
            return;
        },
        enumerable: false,
        configurable: true
    });
    JovoUser.prototype.getPersistableData = function () {
        return {
            data: this.data,
        };
    };
    JovoUser.prototype.setPersistableData = function (data) {
        this.data = (data === null || data === void 0 ? void 0 : data.data) || {};
        return this;
    };
    JovoUser.prototype.toJSON = function () {
        return __assign(__assign({}, this), { jovo: undefined });
    };
    return JovoUser;
}());
export { JovoUser };
