// @jovotech/framework/dist/esm5/JovoSession.js
import { DEFAULT_SESSION_EXPIRES_AFTER_SECONDS } from './index';
var JovoSession = /** @class */ (function () {
    function JovoSession(data) {
        var _a;
        this.id = data === null || data === void 0 ? void 0 : data.id;
        this.data = (data === null || data === void 0 ? void 0 : data.data) || {};
        this.state = data === null || data === void 0 ? void 0 : data.state;
        this.isNew = (_a = data === null || data === void 0 ? void 0 : data.isNew) !== null && _a !== void 0 ? _a : true;
        this.updatedAt = new Date();
        this.createdAt = this.isNew
            ? new Date()
            : (data === null || data === void 0 ? void 0 : data.createdAt)
                ? new Date(data.createdAt)
                : new Date();
    }
    JovoSession.prototype.getPersistableData = function () {
        return {
            id: this.id,
            data: this.data,
            state: this.state,
            createdAt: this.createdAt.toISOString(),
            updatedAt: this.updatedAt.toISOString(),
        };
    };
    JovoSession.prototype.setPersistableData = function (data, config) {
        if (!data) {
            return this;
        }
        var expiresAfterSeconds = (typeof config === 'object' ? config.expiresAfterSeconds : undefined) ||
            DEFAULT_SESSION_EXPIRES_AFTER_SECONDS;
        var expiresAfterMilliseconds = expiresAfterSeconds * 1000;
        var timeDifferenceInMilliseconds = new Date().getTime() - new Date(data.updatedAt).getTime();
        var isExpired = timeDifferenceInMilliseconds > expiresAfterMilliseconds;
        if (isExpired) {
            return this;
        }
        // the loaded session can not be new because it was loaded from the database and is not expired
        this.isNew = false;
        this.id = (data === null || data === void 0 ? void 0 : data.id) || this.id;
        this.data = (data === null || data === void 0 ? void 0 : data.data) || this.data;
        this.state = (data === null || data === void 0 ? void 0 : data.state) || this.state;
        this.updatedAt = new Date();
        this.createdAt = (data === null || data === void 0 ? void 0 : data.createdAt) ? new Date(data.createdAt) : new Date();
        return this;
    };
    return JovoSession;
}());
export { JovoSession };
