// @jovotech/platform-alexa/dist/esm2015/cli/index.js
import { JovoCliPlugin, Log, promptSupportedLocales, } from '@jovotech/cli-core';
import { writeFileSync } from 'fs';
import { join as joinPaths, resolve } from 'path';
import { SupportedLocales } from './constants';
import { BuildHook } from './hooks/BuildHook';
import { DeployHook } from './hooks/DeployHook';
import { GetHook } from './hooks/GetHook';
import { NewHook } from './hooks/NewHook';
import { CertifyCommand } from './commands/CertifyCommand';
import { ValidateCommand } from './commands/ValidateCommand';
export class AlexaCli extends JovoCliPlugin {
    constructor(config) {
        super(config);
        // Convert boolean value of this.config.conversations to object to handle it unified across hooks
        if (typeof this.config.conversations === 'boolean') {
            const { conversations: defaultConversationsConfig } = this.getDefaultConfig();
            this.config.conversations = Object.assign(Object.assign({}, defaultConversationsConfig), { enabled: this.config.conversations });
        }
    }
    get id() {
        return 'alexa';
    }
    get type() {
        return 'platform';
    }
    get platformDirectory() {
        return `${this.type}.${this.id}`;
    }
    getDefaultConfig() {
        return {
            conversations: {
                enabled: false,
                // ! Since getDefaultConfig() is called before this.id is assigned, we need to set it manually
                directory: joinPaths('resources', 'alexa', 'conversations'),
                sessionStartDelegationStrategy: {
                    target: 'skill',
                },
                acdlDirectory: 'acdl',
                responsesDirectory: 'responses',
                skipValidation: false,
            },
        };
    }
    async getInitConfig() {
        const initConfig = {};
        // Since this.getInitConfig() is called when this plugin is added to a new
        // Jovo project, we can add external bundle dependencies here
        const packageJsonPath = resolve(joinPaths(this.$cli.projectPath, this.$context.projectName, 'package.json'));
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        const packageJson = require(packageJsonPath);
        packageJson.scripts.bundle = `${packageJson.scripts.bundle} --external:@alexa/*`;
        writeFileSync(packageJsonPath, JSON.stringify(packageJson, null, 2));
        // Check for invalid locales and provide a default locale map.
        for (const locale of this.$context.locales) {
            if (!SupportedLocales.includes(locale)) {
                // Prompt user for alternative locale.
                Log.spacer();
                const { locales } = await promptSupportedLocales(locale, 'Alexa', SupportedLocales);
                if (!locales.length) {
                    continue;
                }
                if (!initConfig.locales) {
                    initConfig.locales = {};
                }
                initConfig.locales[locale] = locales;
            }
        }
        return initConfig;
    }
    getHooks() {
        return [BuildHook, GetHook, DeployHook, NewHook];
    }
    getCommands() {
        return [CertifyCommand, ValidateCommand];
    }
    /**
     * The base path to platform's build folder
     */
    get platformPath() {
        return joinPaths(this.$cli.project.getBuildPath(), this.platformDirectory);
    }
    get resourcesDirectory() {
        return joinPaths('resources', this.id);
    }
    /**
     * The path to Alexa skill package folder
     */
    get skillPackagePath() {
        return joinPaths(this.platformPath, 'skill-package');
    }
    /**
     * The path to the skill.json file
     */
    get skillJsonPath() {
        return joinPaths(this.skillPackagePath, 'skill.json');
    }
    get modelsPath() {
        return joinPaths(this.skillPackagePath, 'interactionModels', 'custom');
    }
    get accountLinkingPath() {
        return joinPaths(this.skillPackagePath, 'accountLinking.json');
    }
    get askConfigFolderPath() {
        return joinPaths(this.platformPath, '.ask');
    }
    get askConfigPath() {
        return joinPaths(this.askConfigFolderPath, 'ask-states.json');
    }
    get askResourcesPath() {
        return joinPaths(this.platformPath, 'ask-resources.json');
    }
    get conversationsDirectory() {
        return joinPaths(this.skillPackagePath, 'conversations');
    }
    get responseDirectory() {
        return joinPaths(this.skillPackagePath, 'response');
    }
    getModelPath(locale) {
        return joinPaths(this.modelsPath, `${locale}.json`);
    }
}
