// @jovotech/framework/dist/esm2015/plugins/BasicLogging.js
import chalk from 'chalk';
import colorize from 'json-colorizer';
import { LoggingFormat } from '../enums';
import { Plugin } from '../Plugin';
import { copy, mask } from '../utilities';
export class BasicLogging extends Plugin {
    getDefaultConfig() {
        return {
            skipTests: true,
            enabled: true,
            request: {
                enabled: true,
                excludedObjects: [],
                maskedObjects: [],
                objects: [],
            },
            response: {
                enabled: true,
                excludedObjects: [],
                maskedObjects: [],
                objects: [],
            },
            format: LoggingFormat.Pretty,
            styling: true,
            maskValue: '[ Hidden ]',
            indentation: 2,
            colorizeSettings: {
                colors: {
                    STRING_KEY: 'white',
                    STRING_LITERAL: 'green',
                    NUMBER_LITERAL: 'yellow',
                    BRACE: 'white.bold',
                },
            },
        };
    }
    constructor(config) {
        if ((config === null || config === void 0 ? void 0 : config.enabled) === false) {
            if (typeof config.request === 'undefined') {
                config.request = false;
            }
            if (typeof config.response === 'undefined') {
                config.response = false;
            }
        }
        super(config);
        if (typeof this.config.request === 'boolean') {
            this.config.request = {
                objects: [],
                maskedObjects: [],
                excludedObjects: [],
                enabled: this.config.request,
            };
        }
        if (typeof this.config.response === 'boolean') {
            this.config.response = {
                objects: [],
                maskedObjects: [],
                excludedObjects: [],
                enabled: this.config.response,
            };
        }
    }
    mount(parent) {
        parent.middlewareCollection.use('request.start', (jovo) => {
            if (this.config.request.enabled) {
                return this.logRequest(jovo);
            }
        });
        parent.middlewareCollection.use('response.end', (jovo) => {
            if (this.config.response.enabled) {
                return this.logResponse(jovo);
            }
        });
    }
    async logRequest(jovo) {
        jovo.$data._BASIC_LOGGING_START = new Date().getTime();
        const requestConfig = this.config.request;
        const requestCopy = copy(jovo.$request, {
            include: requestConfig.objects,
            exclude: requestConfig.excludedObjects,
        });
        // Mask properties according to configuration
        if (requestConfig.maskedObjects && requestConfig.maskedObjects.length > 0) {
            mask(requestCopy, requestConfig.maskedObjects, this.config.maskValue);
        }
        if (this.config.format === LoggingFormat.Pretty) {
            if (this.config.styling) {
                // eslint-disable-next-line no-console
                console.log(chalk.bgWhite.black('\n\n >>>>> Request - ' + new Date().toISOString() + ' '));
            }
            // eslint-disable-next-line no-console
            console.log(colorize(JSON.stringify(requestCopy, null, this.config.indentation || 2), this.config.colorizeSettings));
        }
        else if (this.config.format === LoggingFormat.Json) {
            // eslint-disable-next-line no-console
            console.log(colorize(JSON.stringify(requestCopy), this.config.colorizeSettings));
        }
    }
    async logResponse(jovo) {
        const basicLoggingEnd = new Date().getTime();
        const duration = jovo.$data._BASIC_LOGGING_START
            ? basicLoggingEnd - jovo.$data._BASIC_LOGGING_START
            : 0;
        const responseConfig = this.config.response;
        const responseCopy = copy(jovo.$response, {
            include: responseConfig.objects,
            exclude: responseConfig.excludedObjects,
        });
        if (responseConfig.maskedObjects && responseConfig.maskedObjects.length > 0) {
            mask(responseCopy, responseConfig.maskedObjects, this.config.maskValue);
        }
        if (this.config.format === LoggingFormat.Pretty) {
            // eslint-disable-next-line no-console
            console.log(chalk.bgGray.white('\n\n <<<<< Response - ' + new Date().toISOString() + ' ') +
                ' ✔️ ' +
                duration +
                'ms');
            // eslint-disable-next-line no-console
            console.log(colorize(JSON.stringify(responseCopy, null, this.config.indentation || 2), this.config.colorizeSettings));
        }
        else if (this.config.format === LoggingFormat.Json) {
            // eslint-disable-next-line no-console
            console.log(colorize(JSON.stringify(responseCopy), this.config.colorizeSettings));
        }
    }
}
