// @jovotech/platform-core/dist/esm5/CoreRequest.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { JovoRequest } from '@jovotech/framework';
var CoreRequest = /** @class */ (function (_super) {
    __extends(CoreRequest, _super);
    function CoreRequest() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CoreRequest.prototype.getLocale = function () {
        return this.locale;
    };
    CoreRequest.prototype.getIntent = function () {
        var _a, _b, _c;
        return ((_a = this.input) === null || _a === void 0 ? void 0 : _a.intent) || ((_c = (_b = this.input) === null || _b === void 0 ? void 0 : _b.nlu) === null || _c === void 0 ? void 0 : _c.intent);
    };
    CoreRequest.prototype.setIntent = function (intent) {
        if (!this.input) {
            this.input = {};
        }
        this.input.intent = intent;
    };
    CoreRequest.prototype.getEntities = function () {
        var _a, _b, _c;
        return ((_a = this.input) === null || _a === void 0 ? void 0 : _a.entities) || ((_c = (_b = this.input) === null || _b === void 0 ? void 0 : _b.nlu) === null || _c === void 0 ? void 0 : _c.entities);
    };
    CoreRequest.prototype.getInputType = function () {
        var _a;
        return (_a = this.input) === null || _a === void 0 ? void 0 : _a.type;
    };
    CoreRequest.prototype.getInputText = function () {
        var _a;
        return (_a = this.input) === null || _a === void 0 ? void 0 : _a.text;
    };
    CoreRequest.prototype.getInputAudio = function () {
        var _a;
        return (_a = this.input) === null || _a === void 0 ? void 0 : _a.audio;
    };
    CoreRequest.prototype.getSessionData = function () {
        var _a;
        return (_a = this.context) === null || _a === void 0 ? void 0 : _a.session;
    };
    CoreRequest.prototype.getSessionId = function () {
        var _a, _b;
        return (_b = (_a = this.context) === null || _a === void 0 ? void 0 : _a.session) === null || _b === void 0 ? void 0 : _b.id;
    };
    CoreRequest.prototype.isNewSession = function () {
        var _a, _b;
        return (_b = (_a = this.context) === null || _a === void 0 ? void 0 : _a.session) === null || _b === void 0 ? void 0 : _b.isNew;
    };
    CoreRequest.prototype.getDeviceCapabilities = function () {
        var _a, _b;
        return (_b = (_a = this.context) === null || _a === void 0 ? void 0 : _a.device) === null || _b === void 0 ? void 0 : _b.capabilities;
    };
    CoreRequest.prototype.setLocale = function (locale) {
        this.locale = locale;
    };
    CoreRequest.prototype.setSessionData = function (data) {
        var _a, _b;
        if (!((_b = (_a = this.context) === null || _a === void 0 ? void 0 : _a.session) === null || _b === void 0 ? void 0 : _b.data)) {
            return;
        }
        this.context.session.data = data;
    };
    CoreRequest.prototype.getUserId = function () {
        var _a;
        return (_a = this.context) === null || _a === void 0 ? void 0 : _a.user.id;
    };
    CoreRequest.prototype.setUserId = function (userId) {
        if (!this.context) {
            // TODO: What to do here?
            return;
        }
        this.context.user.id = userId;
    };
    CoreRequest.prototype.getRequestId = function () {
        return this.id;
    };
    return CoreRequest;
}(JovoRequest));
export { CoreRequest };
