// @jovotech/platform-alexa/dist/esm2015/output/templates/ConnectionScheduleFoodEstablishmentReservationOutput.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { BaseOutput, Output } from '@jovotech/framework';
import { OnCompletion } from '../models/common/OnCompletion';
import { StartConnectionOutput } from './StartConnectionOutput';
let ConnectionScheduleFoodEstablishmentReservationOutput = class ConnectionScheduleFoodEstablishmentReservationOutput extends BaseOutput {
    getDefaultOptions() {
        return {
            onCompletion: OnCompletion.ResumeSession,
            restaurant: {
                '@type': 'Restaurant',
                '@version': '1',
                'name': '',
                'location': {
                    '@type': 'PostalAddress',
                    '@version': '1',
                    'streetAddress': '',
                    'locality': '',
                    'region': '',
                    'postalCode': '',
                },
            },
        };
    }
    build() {
        return new StartConnectionOutput(this.jovo, {
            taskName: { name: 'ScheduleFoodEstablishmentReservation', amazonPredefinedTask: true },
            taskVersion: 1,
            shouldEndSession: this.options.shouldEndSession,
            onCompletion: this.options.onCompletion,
            input: {
                '@type': 'ScheduleFoodEstablishmentReservationRequest',
                '@version': '1',
                'startTime': this.options.startTime,
                'partySize': this.options.partySize,
                'restaurant': this.options.restaurant,
            },
            token: this.options.token,
        }).build();
    }
};
ConnectionScheduleFoodEstablishmentReservationOutput = __decorate([
    Output()
], ConnectionScheduleFoodEstablishmentReservationOutput);
export { ConnectionScheduleFoodEstablishmentReservationOutput };
