// @jovotech/platform-alexa/dist/esm2015/output/templates/ConnectionVerifyPersonOutput.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { BaseOutput, Output } from '@jovotech/framework';
import { OnCompletion } from '../models/common/OnCompletion';
import { StartConnectionOutput } from './StartConnectionOutput';
export var PolicyName;
(function (PolicyName) {
    PolicyName["VoicePin"] = "VOICE_PIN";
})(PolicyName || (PolicyName = {}));
let ConnectionVerifyPersonOutput = class ConnectionVerifyPersonOutput extends BaseOutput {
    getDefaultOptions() {
        return {
            onCompletion: OnCompletion.ResumeSession,
            level: 400,
            policyName: PolicyName.VoicePin,
        };
    }
    build() {
        return new StartConnectionOutput(this.jovo, {
            taskName: { name: 'VerifyPerson', amazonPredefinedTask: true },
            taskVersion: 2,
            shouldEndSession: this.options.shouldEndSession,
            onCompletion: this.options.onCompletion,
            input: {
                requestedAuthenticationConfidenceLevel: {
                    level: this.options.level,
                    customPolicy: {
                        policyName: this.options.policyName,
                    },
                },
            },
            token: this.options.token,
        }).build();
    }
};
ConnectionVerifyPersonOutput = __decorate([
    Output()
], ConnectionVerifyPersonOutput);
export { ConnectionVerifyPersonOutput };
