// @jovotech/platform-alexa/dist/esm2015/output/templates/ConnectionPrintWebPageOutput.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { BaseOutput, Output } from '@jovotech/framework';
import { OnCompletion } from '../models/common/OnCompletion';
import { StartConnectionOutput } from './StartConnectionOutput';
let ConnectionPrintWebPageOutput = class ConnectionPrintWebPageOutput extends BaseOutput {
    getDefaultOptions() {
        return {
            onCompletion: OnCompletion.ResumeSession,
            title: '',
            url: '',
        };
    }
    build() {
        return new StartConnectionOutput(this.jovo, {
            taskName: { name: 'PrintWebPage', amazonPredefinedTask: true },
            taskVersion: 1,
            shouldEndSession: this.options.shouldEndSession,
            onCompletion: this.options.onCompletion,
            input: {
                '@type': 'PrintWebPageRequest',
                '@version': '1',
                'title': this.options.title,
                'description': this.options.description,
                'url': this.options.url,
            },
            token: this.options.token,
        }).build();
    }
};
ConnectionPrintWebPageOutput = __decorate([
    Output()
], ConnectionPrintWebPageOutput);
export { ConnectionPrintWebPageOutput };
