// @jovotech/platform-alexa/dist/esm2015/output/templates/LinkAccountCardOutput.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { BaseOutput, Output } from '@jovotech/framework';
let LinkAccountCardOutput = class LinkAccountCardOutput extends BaseOutput {
    build() {
        return {
            message: this.options.message,
            listen: false,
            platforms: {
                alexa: {
                    nativeResponse: {
                        response: {
                            card: {
                                type: 'LinkAccount',
                            },
                        },
                    },
                },
            },
        };
    }
};
LinkAccountCardOutput = __decorate([
    Output()
], LinkAccountCardOutput);
export { LinkAccountCardOutput };
