// @jovotech/platform-alexa/dist/esm2015/output/templates/VideoAppLaunchOutput.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { Output, BaseOutput } from '@jovotech/framework';
let VideoAppLaunchOutput = class VideoAppLaunchOutput extends BaseOutput {
    async build() {
        const videoAppLaunchDirective = {
            type: 'VideoApp.Launch',
            videoItem: this.options.videoItem,
        };
        return {
            platforms: {
                alexa: {
                    nativeResponse: {
                        response: {
                            directives: [videoAppLaunchDirective],
                            // Responses containing a VideoApp.Launch Directive must not have a shouldEndSession parameter
                            shouldEndSession: undefined,
                        },
                    },
                },
            },
        };
    }
};
VideoAppLaunchOutput = __decorate([
    Output()
], VideoAppLaunchOutput);
export { VideoAppLaunchOutput };
