// @jovotech/framework/dist/esm2015/I18Next.js
import i18next from 'i18next';
import { Plugin } from './Plugin';
export class I18Next extends Plugin {
    constructor() {
        super(...arguments);
        this.i18n = i18next;
    }
    getDefaultConfig() {
        return {
            interpolation: {
                escapeValue: false,
                skipOnVariables: false, // Added for backwards compatibility, @see https://www.i18next.com/misc/migration-guide#skiponvariables-true
            },
            returnObjects: true,
            compatibilityJSON: 'v3', // Added for backwards compatibility, @see https://www.i18next.com/misc/migration-guide#json-format-v4-pluralization
        };
    }
    async initialize() {
        await this.i18n.init(this.config);
    }
    t(path, options) {
        if (options === null || options === void 0 ? void 0 : options.platform) {
            if (Array.isArray(path)) {
                for (const p of path) {
                    const platformPath = `${options.platform}:translation:${p}`;
                    if (this.i18n.exists(platformPath, options)) {
                        return this.i18n.t(platformPath, options);
                    }
                }
            }
            else {
                const platformPath = `${options.platform}:translation:${path}`;
                if (this.i18n.exists(platformPath, options)) {
                    return this.i18n.t(platformPath, options);
                }
            }
        }
        return this.i18n.t(path, options);
    }
}
