// @jovotech/platform-googleassistant/dist/esm5/output/models/common/Image.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { IsInt, IsNotEmpty, IsOptional, IsString, IsUrl } from '@jovotech/output';
export var ImageFill;
(function (ImageFill) {
    ImageFill["Unspecified"] = "UNSPECIFIED";
    ImageFill["Gray"] = "GRAY";
    ImageFill["White"] = "WHITE";
    ImageFill["Cropped"] = "CROPPED";
})(ImageFill || (ImageFill = {}));
var Image = /** @class */ (function () {
    function Image() {
    }
    __decorate([
        IsUrl({ protocols: ['https', 'http'] }),
        __metadata("design:type", String)
    ], Image.prototype, "url", void 0);
    __decorate([
        IsString(),
        IsNotEmpty(),
        __metadata("design:type", String)
    ], Image.prototype, "alt", void 0);
    __decorate([
        IsOptional(),
        IsInt(),
        __metadata("design:type", Number)
    ], Image.prototype, "height", void 0);
    __decorate([
        IsOptional(),
        IsInt(),
        __metadata("design:type", Number)
    ], Image.prototype, "width", void 0);
    return Image;
}());
export { Image };
