// @jovotech/platform-alexa/dist/esm5/output/models/card/Card.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { IsOptional, IsArray, IsEnum, IsNotEmpty, IsString, Type, } from '@jovotech/output';
import { IsValidCardImage } from '../../decorators/validation/IsValidCardImage';
import { IsValidCardString } from '../../decorators/validation/IsValidCardString';
import { CardImage } from './CardImage';
export var CardType;
(function (CardType) {
    CardType["Simple"] = "Simple";
    CardType["Standard"] = "Standard";
    CardType["LinkAccount"] = "LinkAccount";
    CardType["AskForPermissionsConsent"] = "AskForPermissionsConsent";
})(CardType || (CardType = {}));
var Card = /** @class */ (function () {
    function Card() {
    }
    Card.prototype.toCard = function () {
        var _a, _b;
        var card = {
            title: (this.title || ''),
        };
        if (this.text || this.content) {
            card.content = this.text || this.content;
        }
        if (((_a = this.image) === null || _a === void 0 ? void 0 : _a.largeImageUrl) || ((_b = this.image) === null || _b === void 0 ? void 0 : _b.smallImageUrl)) {
            card.imageUrl = this.image.largeImageUrl || this.image.smallImageUrl;
        }
        return card;
    };
    __decorate([
        IsEnum(CardType),
        __metadata("design:type", String)
    ], Card.prototype, "type", void 0);
    __decorate([
        IsValidCardString([CardType.Simple, CardType.Standard, CardType.AskForPermissionsConsent]),
        __metadata("design:type", Object)
    ], Card.prototype, "title", void 0);
    __decorate([
        IsValidCardString([CardType.Simple, CardType.AskForPermissionsConsent]),
        __metadata("design:type", Object)
    ], Card.prototype, "content", void 0);
    __decorate([
        IsValidCardString([CardType.Standard, CardType.AskForPermissionsConsent]),
        __metadata("design:type", Object)
    ], Card.prototype, "text", void 0);
    __decorate([
        IsValidCardImage([CardType.Standard]),
        Type(function () { return CardImage; }),
        __metadata("design:type", Object)
    ], Card.prototype, "image", void 0);
    __decorate([
        IsOptional(),
        IsArray(),
        IsString({ each: true }),
        IsNotEmpty({ each: true }),
        __metadata("design:type", Array)
    ], Card.prototype, "permissions", void 0);
    return Card;
}());
export { Card };
