// @jovotech/cli-core/dist/constants.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BULB = exports.CONSTRUCTION = exports.DISK = exports.MAGNIFYING_GLASS = exports.DOWNLOAD = exports.PACKAGE = exports.ARROW_UP = exports.REVERSE_ARROWS = exports.SPARKLES = exports.ERROR_PREFIX = exports.ERROR = exports.SUCCESS = exports.WARNING = exports.ROCKET = exports.WRENCH = exports.CRYSTAL_BALL = exports.CLOUD = exports.STATION = exports.STAR = exports.TADA = exports.OK_HAND = exports.ANSWER_BACKUP = exports.ANSWER_CANCEL = exports.ANSWER_UPDATE = exports.ANSWER_OVERWRITE = exports.SUPPORTED_LANGUAGES = exports.REPO_URL = exports.JOVO_WEBHOOK_URL = exports.DEFAULT_LOCALE = void 0;
const os = __importStar(require("os"));
const node_emoji_1 = require("node-emoji");
const chalk_1 = __importDefault(require("chalk"));
const is_ci_1 = __importDefault(require("is-ci"));
exports.DEFAULT_LOCALE = 'en';
exports.JOVO_WEBHOOK_URL = 'https://webhook.jovo.cloud';
exports.REPO_URL = 'https://www.jovo.tech/repo/sample-apps';
exports.SUPPORTED_LANGUAGES = ['typescript', 'javascript'];
// ####### PROMPT ANSWERS #######
exports.ANSWER_OVERWRITE = 'overwrite';
exports.ANSWER_UPDATE = 'update';
exports.ANSWER_CANCEL = 'cancel';
exports.ANSWER_BACKUP = 'backup';
// ####### EMOJIS #######
// Check if current shell supports emoji. On Windows, this is only the case for Windows Terminal.
// Available emoji can be found here: https://www.webfx.com/tools/emoji-cheat-sheet
const SUPPORTS_EMOJI = !is_ci_1.default && (os.platform() === 'win32' ? !!process.env.WT_SESSION : true);
exports.OK_HAND = SUPPORTS_EMOJI ? node_emoji_1.get('ok_hand') : '';
exports.TADA = SUPPORTS_EMOJI ? node_emoji_1.get('tada') : '';
exports.STAR = SUPPORTS_EMOJI ? node_emoji_1.get('star') : '';
exports.STATION = SUPPORTS_EMOJI ? node_emoji_1.get('station') : '';
exports.CLOUD = SUPPORTS_EMOJI ? node_emoji_1.get('cloud') : '';
exports.CRYSTAL_BALL = SUPPORTS_EMOJI ? node_emoji_1.get('crystal_ball') : '';
exports.WRENCH = SUPPORTS_EMOJI ? node_emoji_1.get('wrench') : '';
exports.ROCKET = SUPPORTS_EMOJI ? node_emoji_1.get('rocket') : '';
exports.WARNING = chalk_1.default.yellow('WARN');
exports.SUCCESS = SUPPORTS_EMOJI ? chalk_1.default.green('✔') : chalk_1.default.green('√');
exports.ERROR = chalk_1.default.red('x');
exports.ERROR_PREFIX = chalk_1.default.red('›');
exports.SPARKLES = SUPPORTS_EMOJI ? node_emoji_1.get('sparkles') : '';
exports.REVERSE_ARROWS = SUPPORTS_EMOJI ? node_emoji_1.get('leftwards_arrow_with_hook:') : '';
exports.ARROW_UP = SUPPORTS_EMOJI ? node_emoji_1.get('arrow_up') : '';
exports.PACKAGE = SUPPORTS_EMOJI ? node_emoji_1.get('package') : '';
exports.DOWNLOAD = SUPPORTS_EMOJI ? node_emoji_1.get('inbox_tray') : '';
exports.MAGNIFYING_GLASS = SUPPORTS_EMOJI ? node_emoji_1.get('mag') : '';
exports.DISK = SUPPORTS_EMOJI ? node_emoji_1.get('floppy_disk') : '';
exports.CONSTRUCTION = SUPPORTS_EMOJI ? node_emoji_1.get('construction') : '';
exports.BULB = SUPPORTS_EMOJI ? node_emoji_1.get('bulb') : '';
