// @jovotech/cli-core/dist/PluginComponent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginComponent = void 0;
class PluginComponent {
    static install(cli, plugin, emitter) {
        this.prototype.$cli = cli;
        this.prototype.$plugin = plugin;
        this.prototype.$emitter = emitter;
        // Load action set.
        this.prototype.install();
        // Register events to emitter.
        this.prototype.loadMiddlewareCollection();
    }
    install() { }
    loadMiddlewareCollection() {
        for (const [key, value] of Object.entries(this.middlewareCollection || {})) {
            if (!value) {
                continue;
            }
            for (const fn of value) {
                this.$emitter.on(key, fn);
            }
        }
    }
    uninstall() {
        for (const [key, value] of Object.entries(this.middlewareCollection || {})) {
            if (!value) {
                continue;
            }
            for (const fn of value) {
                this.$emitter.off(key, fn);
            }
        }
    }
}
exports.PluginComponent = PluginComponent;
