// @jovotech/cli-core/dist/prints.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.printPackages = exports.printUserInput = exports.printAskProfile = exports.printSubHeadline = exports.printHighlight = exports.printStage = void 0;
const chalk_1 = __importDefault(require("chalk"));
const indent_string_1 = __importDefault(require("indent-string"));
/**
 * Returns a string of a stage in white bold font.
 * @param stage - Stage to print.
 */
function printStage(stage) {
    return stage ? `[stage: ${printHighlight(stage)}]` : '';
}
exports.printStage = printStage;
/**
 * Returns a string of a white, bold text.
 * @param text - Text to highlight.
 */
function printHighlight(text) {
    return chalk_1.default.white.bold(text);
}
exports.printHighlight = printHighlight;
/**
 * Returns a string of a grey sub headline.
 * @param hl - Subheadline to print.
 */
function printSubHeadline(hl) {
    return chalk_1.default.white.dim(indent_string_1.default(`>> ${hl}\n`, 2));
}
exports.printSubHeadline = printSubHeadline;
/**
 * Returns a string of an ASK profile in white bold font.
 * @param askProfile - Optional ASK profile to print.
 */
function printAskProfile(askProfile) {
    return askProfile ? `[ASK profile: ${printHighlight(askProfile)}]` : '';
}
exports.printAskProfile = printAskProfile;
/**
 * Prints user input in prompts in a bright blue.
 * @param input - User input to color in.
 */
function printUserInput(input) {
    return chalk_1.default.blueBright(input);
}
exports.printUserInput = printUserInput;
function printPackages(packages) {
    const output = [];
    if (packages.length) {
        for (const pkg of packages) {
            output.push(`  - ${pkg.name}: ${pkg.version.local} ${pkg.isDev ? chalk_1.default.grey('[dev]') : ''} ${pkg.version.npm !== pkg.version.local ? chalk_1.default.green(`-> ${pkg.version.npm}`) : ''}`);
        }
    }
    return output.join('\n');
}
exports.printPackages = printPackages;
