// @jovotech/platform-googleassistant/dist/esm5/output/models/content/Media.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { IsArray, IsEnum, IsNotEmpty, IsOptional, IsString, IsUrl, Type, ValidateNested, } from '@jovotech/output';
import { IsValidMediaObjectImage } from '../../decorators/validation/IsValidMediaObjectImage';
import { Image } from '../common/Image';
export var MediaType;
(function (MediaType) {
    MediaType["Unspecified"] = "MEDIA_TYPE_UNSPECIFIED";
    MediaType["Audio"] = "AUDIO";
    MediaType["MediaStatusAck"] = "MEDIA_STATUS_ACK";
})(MediaType || (MediaType = {}));
export var OptionalMediaControls;
(function (OptionalMediaControls) {
    OptionalMediaControls["Unspecified"] = "OPTIONAL_MEDIA_CONTROLS_UNSPECIFIED";
    OptionalMediaControls["Paused"] = "PAUSED";
    OptionalMediaControls["Stopped"] = "STOPPED";
})(OptionalMediaControls || (OptionalMediaControls = {}));
var MediaImage = /** @class */ (function () {
    function MediaImage() {
    }
    __decorate([
        IsValidMediaObjectImage(),
        Type(function () { return Image; }),
        __metadata("design:type", Image)
    ], MediaImage.prototype, "large", void 0);
    __decorate([
        IsValidMediaObjectImage(),
        Type(function () { return Image; }),
        __metadata("design:type", Image)
    ], MediaImage.prototype, "icon", void 0);
    return MediaImage;
}());
export { MediaImage };
var MediaObject = /** @class */ (function () {
    function MediaObject() {
    }
    __decorate([
        IsString(),
        IsNotEmpty(),
        __metadata("design:type", String)
    ], MediaObject.prototype, "name", void 0);
    __decorate([
        IsString(),
        __metadata("design:type", String)
    ], MediaObject.prototype, "description", void 0);
    __decorate([
        IsUrl({ protocols: ['https', 'http'] }),
        __metadata("design:type", String)
    ], MediaObject.prototype, "url", void 0);
    __decorate([
        ValidateNested(),
        Type(function () { return MediaImage; }),
        __metadata("design:type", MediaImage)
    ], MediaObject.prototype, "image", void 0);
    return MediaObject;
}());
export { MediaObject };
var Media = /** @class */ (function () {
    function Media() {
    }
    __decorate([
        IsEnum(MediaType),
        __metadata("design:type", String)
    ], Media.prototype, "mediaType", void 0);
    __decorate([
        IsString(),
        IsNotEmpty(),
        __metadata("design:type", String)
    ], Media.prototype, "startOffset", void 0);
    __decorate([
        IsOptional(),
        IsArray(),
        IsEnum(OptionalMediaControls, { each: true }),
        __metadata("design:type", Array)
    ], Media.prototype, "optionalMediaControls", void 0);
    __decorate([
        IsArray(),
        ValidateNested({ each: true }),
        Type(function () { return MediaObject; }),
        __metadata("design:type", Array)
    ], Media.prototype, "mediaObjects", void 0);
    return Media;
}());
export { Media };
