// @jovotech/output/dist/esm5/OutputTemplateConverterStrategy.js
import { SsmlUtilities } from '@jovotech/common';
import { NormalizedOutputTemplate, plainToClass, } from '.';
import { OutputHelpers } from './OutputHelpers';
import _merge from 'lodash.merge';
var OutputTemplateConverterStrategy = /** @class */ (function () {
    function OutputTemplateConverterStrategy(config) {
        this.config = _merge(this.getDefaultConfig(), config || {});
    }
    OutputTemplateConverterStrategy.prototype.getDefaultConfig = function () {
        return {
            omitWarnings: false,
            sanitization: true,
            validation: true,
        };
    };
    // Normalize the output:
    // 1. get the platform specific output
    // 2. get the randomized output
    // 3. return instance(s) of NormalizedOutputTemplate
    OutputTemplateConverterStrategy.prototype.normalizeOutput = function (output) {
        var _this = this;
        var normalize = function (outputTemplate) {
            return _this.getRandomizedOutput(_this.getPlatformSpecificOutput(outputTemplate));
        };
        var normalizedOutput = Array.isArray(output) ? output.map(normalize) : normalize(output);
        return plainToClass(NormalizedOutputTemplate, normalizedOutput);
    };
    // Normalize the response:
    // - return instance of responseClass
    OutputTemplateConverterStrategy.prototype.normalizeResponse = function (response) {
        return plainToClass(this.responseClass, response);
    };
    OutputTemplateConverterStrategy.prototype.getRandomizedOutput = function (output) {
        if (Array.isArray(output.message)) {
            output.message = OutputHelpers.randomize(output.message);
        }
        if (Array.isArray(output.reprompt)) {
            output.reprompt = OutputHelpers.randomize(output.reprompt);
        }
        return output;
    };
    OutputTemplateConverterStrategy.prototype.getPlatformSpecificOutput = function (output) {
        var _this = this;
        return NormalizedOutputTemplate.getKeys().reduce(function (outputCopy, outputKey) {
            var _a, _b;
            if (outputKey === 'platforms') {
                // remove the platforms-output of all other platforms due to not being used anyways
                if ((_a = output.platforms) === null || _a === void 0 ? void 0 : _a[_this.platformName]) {
                    outputCopy.platforms = {};
                    outputCopy.platforms[_this.platformName] = (_b = output.platforms) === null || _b === void 0 ? void 0 : _b[_this.platformName];
                }
                return outputCopy;
            }
            var newValue = _this.getOutputValue(output, outputKey);
            if (typeof newValue !== 'undefined') {
                outputCopy[outputKey] = newValue;
            }
            return outputCopy;
        }, {});
    };
    OutputTemplateConverterStrategy.prototype.getOutputValue = function (output, key) {
        var _a, _b;
        var platformValue = (_b = (_a = output.platforms) === null || _a === void 0 ? void 0 : _a[this.platformName]) === null || _b === void 0 ? void 0 : _b[key];
        if (platformValue === null) {
            return undefined;
        }
        return platformValue !== null && platformValue !== void 0 ? platformValue : output[key];
    };
    OutputTemplateConverterStrategy.prototype.shouldSanitize = function (rule) {
        if (!rule) {
            return !!this.config.sanitization;
        }
        return typeof this.config.sanitization === 'object'
            ? this.config.sanitization[rule]
            : this.config.sanitization;
    };
    OutputTemplateConverterStrategy.prototype.sanitizeMessage = function (message, path, maxLength, offset) {
        var _a, _b;
        if (offset === void 0) { offset = 0; }
        var speechMaxLength = (typeof maxLength === 'number' ? maxLength : maxLength.speech || Infinity) - offset;
        var textMaxLength = typeof maxLength === 'number' ? maxLength : maxLength.text || Infinity;
        var speechLength = typeof message === 'string' ? message.length : ((_a = message.speech) === null || _a === void 0 ? void 0 : _a.length) || 0;
        var textLength = typeof message === 'string' ? message.length : ((_b = message.text) === null || _b === void 0 ? void 0 : _b.length) || 0;
        var isSpeechExceeding = speechLength > speechMaxLength;
        var isTextExceeding = textLength > textMaxLength;
        var isExceeding = isSpeechExceeding || isTextExceeding;
        if (!this.shouldSanitize('trimStrings') || !isExceeding) {
            return message;
        }
        if (typeof message === 'object') {
            if (message.speech && isSpeechExceeding) {
                message.speech = message.speech.slice(0, speechMaxLength);
                this.logStringTrimWarning(path + ".speech", speechMaxLength);
            }
            if (message.text && isTextExceeding) {
                message.text = message.text.slice(0, textMaxLength);
                this.logStringTrimWarning(path + ".text", textMaxLength);
            }
        }
        else {
            var maxLength_1 = SsmlUtilities.isSSML(message) ? speechMaxLength : textMaxLength;
            message = message.slice(0, maxLength_1);
            this.logStringTrimWarning(path, maxLength_1);
        }
        return message;
    };
    OutputTemplateConverterStrategy.prototype.sanitizeDynamicEntities = function (dynamicEntities, path, maxEntries) {
        if (!this.shouldSanitize('trimMaps') ||
            !(dynamicEntities === null || dynamicEntities === void 0 ? void 0 : dynamicEntities.types) ||
            Object.keys(dynamicEntities.types).length <= maxEntries) {
            return dynamicEntities;
        }
        dynamicEntities.types = Object.keys(dynamicEntities)
            .slice(0, maxEntries)
            .reduce(function (map, entityKey) {
            if (!dynamicEntities.types) {
                return map;
            }
            map[entityKey] = dynamicEntities.types[entityKey];
            return map;
        }, {});
        this.logMapTrimWarning(path, maxEntries);
        return dynamicEntities;
    };
    OutputTemplateConverterStrategy.prototype.sanitizeQuickReplies = function (quickReplies, path, maxSize, maxLength) {
        var _this = this;
        if (!this.shouldSanitize('trimArrays') || quickReplies.length <= maxSize) {
            return quickReplies;
        }
        quickReplies = quickReplies.slice(0, maxSize);
        this.logArrayTrimWarning(path, maxSize);
        if (!this.shouldSanitize('trimStrings')) {
            return quickReplies;
        }
        return quickReplies.map(function (quickReply, index) {
            var quickReplyTextLength = typeof quickReply === 'string' ? quickReply.length : quickReply.text.length;
            if (quickReplyTextLength <= maxLength) {
                return quickReply;
            }
            if (typeof quickReply === 'object') {
                quickReply.text = quickReply.text.slice(0, maxLength);
            }
            else {
                quickReply = quickReply.slice(0, maxLength);
            }
            _this.logStringTrimWarning(path + "[" + index + "]", maxLength);
            return quickReply;
        });
    };
    OutputTemplateConverterStrategy.prototype.sanitizeCarousel = function (carousel, path, minSize, maxSize) {
        if (!this.shouldSanitize('trimArrays') ||
            (carousel.items.length >= minSize && carousel.items.length <= maxSize)) {
            return carousel;
        }
        carousel.items = carousel.items.slice(0, maxSize);
        this.logArrayTrimWarning(path, maxSize);
        return carousel;
    };
    OutputTemplateConverterStrategy.prototype.logSanitizationWarning = function (message) {
        if (this.config.omitWarnings) {
            return;
        }
        // eslint-disable-next-line no-console
        console.warn(message);
    };
    OutputTemplateConverterStrategy.prototype.logStringTrimWarning = function (path, maxLength) {
        return this.logSanitizationWarning(this.getTrimMessage(path, maxLength, 'characters'));
    };
    OutputTemplateConverterStrategy.prototype.logArrayTrimWarning = function (path, maxSize) {
        return this.logSanitizationWarning(this.getTrimMessage(path, maxSize, 'items'));
    };
    OutputTemplateConverterStrategy.prototype.logMapTrimWarning = function (path, maxEntries) {
        return this.logSanitizationWarning(this.getTrimMessage(path, maxEntries, 'entries'));
    };
    OutputTemplateConverterStrategy.prototype.getTrimMessage = function (path, max, suffix) {
        var pathStartsWithIndex = /^\[[\d+]].*/.test(path);
        var rootPath = "$output" + (pathStartsWithIndex ? '' : '.');
        path = rootPath + path;
        return path + " was trimmed due to exceeding the limit of " + max + " " + suffix + ".";
    };
    return OutputTemplateConverterStrategy;
}());
export { OutputTemplateConverterStrategy };
