// @jovotech/output/dist/esm5/OutputTemplateConverter.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { plainToClass } from 'class-transformer';
import { validate } from 'class-validator';
import { NormalizedOutputTemplate, } from '.';
import { OutputValidationError } from './errors/OutputValidationError';
var OutputTemplateConverter = /** @class */ (function () {
    function OutputTemplateConverter(strategy) {
        this.strategy = strategy;
    }
    OutputTemplateConverter.prototype.validateOutput = function (output) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.validate(output, NormalizedOutputTemplate)];
            });
        });
    };
    OutputTemplateConverter.prototype.validateResponse = function (response) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.validate(response, this.strategy.responseClass)];
            });
        });
    };
    OutputTemplateConverter.prototype.toResponse = function (output) {
        return __awaiter(this, void 0, void 0, function () {
            var normalizedOutput, errors, response, errors;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        normalizedOutput = this.strategy.normalizeOutput(output);
                        if (!this.shouldValidate('before')) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.validateOutput(normalizedOutput)];
                    case 1:
                        errors = _a.sent();
                        if (errors.length) {
                            throw new OutputValidationError(errors, 'Can not convert.\n');
                        }
                        _a.label = 2;
                    case 2:
                        response = this.strategy.toResponse(normalizedOutput);
                        if (!this.shouldValidate('after')) return [3 /*break*/, 4];
                        return [4 /*yield*/, this.validateResponse(response)];
                    case 3:
                        errors = _a.sent();
                        if (errors.length) {
                            throw new OutputValidationError(errors, 'Conversion caused invalid response.\n');
                        }
                        _a.label = 4;
                    case 4: return [2 /*return*/, response];
                }
            });
        });
    };
    OutputTemplateConverter.prototype.fromResponse = function (response) {
        return __awaiter(this, void 0, void 0, function () {
            var responseInstance, errors, output, errors;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        responseInstance = this.strategy.normalizeResponse(response);
                        if (!this.shouldValidate('before')) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.validateResponse(responseInstance)];
                    case 1:
                        errors = _a.sent();
                        if (errors.length) {
                            throw new OutputValidationError(errors, 'Can not parse.\n');
                        }
                        _a.label = 2;
                    case 2:
                        output = this.strategy.fromResponse(responseInstance);
                        if (!this.shouldValidate('after')) return [3 /*break*/, 4];
                        return [4 /*yield*/, this.validateOutput(output)];
                    case 3:
                        errors = _a.sent();
                        if (errors.length) {
                            throw new OutputValidationError(errors, 'Conversion caused invalid output.\n');
                        }
                        _a.label = 4;
                    case 4: return [2 /*return*/, output];
                }
            });
        });
    };
    OutputTemplateConverter.prototype.shouldValidate = function (key) {
        if (!key) {
            return !!this.strategy.config.validation;
        }
        return typeof this.strategy.config.validation === 'object'
            ? this.strategy.config.validation[key]
            : this.strategy.config.validation;
    };
    OutputTemplateConverter.prototype.validate = function (objOrArray, targetClass) {
        return __awaiter(this, void 0, void 0, function () {
            var getInstance, errorMatrix;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        getInstance = function (item) { return plainToClass(targetClass, item); };
                        if (!Array.isArray(objOrArray)) return [3 /*break*/, 2];
                        return [4 /*yield*/, Promise.all(objOrArray.map(function (item) { return validate(getInstance(item)); }))];
                    case 1:
                        errorMatrix = _a.sent();
                        // TODO: maybe modify key or something to indicate better which item was invalid
                        return [2 /*return*/, errorMatrix.reduce(function (acc, curr) {
                                acc.push.apply(acc, curr);
                                return acc;
                            }, [])];
                    case 2: return [2 /*return*/, validate(getInstance(objOrArray))];
                }
            });
        });
    };
    return OutputTemplateConverter;
}());
export { OutputTemplateConverter };
