// @jovotech/framework/dist/esm2015/testsuite/TestRequest.js
import { InputType, } from '@jovotech/common';
import { JovoRequest, JovoSession } from '..';
export class TestRequest extends JovoRequest {
    constructor() {
        super(...arguments);
        this.isTestRequest = true;
        this.session = new JovoSession({ state: [] });
    }
    getLocale() {
        return this.locale;
    }
    setLocale(locale) {
        this.locale = locale;
    }
    getIntent() {
        return this.intent;
    }
    setIntent(intent) {
        this.intent = intent;
    }
    getEntities() {
        return;
    }
    getInputType() {
        return this.session.isNew ? InputType.Launch : InputType.Intent;
    }
    getInputText() {
        return;
    }
    getInputAudio() {
        return;
    }
    setSessionData(data) {
        this.session.data = data;
    }
    getUserId() {
        return this.userId;
    }
    setUserId(userId) {
        this.userId = userId;
    }
    getSessionData() {
        return this.session.data;
    }
    getSessionId() {
        return this.session.id;
    }
    isNewSession() {
        return this.session.isNew;
    }
    getDeviceCapabilities() {
        return;
    }
    getRequestId() {
        return;
    }
}
