// @jovotech/platform-alexa/dist/esm2015/output/models/common/OutputSpeech.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { IsEnum, IsOptional } from '@jovotech/output';
import { IsValidOutputSpeechString } from '../../decorators/validation/IsValidOutputSpeechString';
export var OutputSpeechType;
(function (OutputSpeechType) {
    OutputSpeechType["Plain"] = "PlainText";
    OutputSpeechType["Ssml"] = "SSML";
})(OutputSpeechType || (OutputSpeechType = {}));
export var PlayBehavior;
(function (PlayBehavior) {
    PlayBehavior["Enqueue"] = "ENQUEUE";
    PlayBehavior["ReplaceAll"] = "REPLACE_ALL";
    PlayBehavior["ReplaceEnqueued"] = "REPLACE_ENQUEUED";
})(PlayBehavior || (PlayBehavior = {}));
export class OutputSpeech {
    toMessage() {
        if (this.type === OutputSpeechType.Ssml && this.ssml) {
            return {
                speech: this.ssml,
            };
        }
        if (this.type === OutputSpeechType.Plain && this.text) {
            return {
                text: this.text,
            };
        }
        return '';
    }
}
__decorate([
    IsEnum(OutputSpeechType),
    __metadata("design:type", String)
], OutputSpeech.prototype, "type", void 0);
__decorate([
    IsValidOutputSpeechString(OutputSpeechType.Plain),
    __metadata("design:type", Object)
], OutputSpeech.prototype, "text", void 0);
__decorate([
    IsValidOutputSpeechString(OutputSpeechType.Ssml),
    __metadata("design:type", Object)
], OutputSpeech.prototype, "ssml", void 0);
__decorate([
    IsOptional(),
    IsEnum(PlayBehavior),
    __metadata("design:type", String)
], OutputSpeech.prototype, "playBehavior", void 0);
