// @jovotech/db-filedb/dist/esm2015/FileDb.js
import { DbPlugin } from '@jovotech/framework';
import fs from 'fs';
import path from 'path';
import process from 'process';
export class FileDb extends DbPlugin {
    constructor(config) {
        super(config);
    }
    getDefaultConfig() {
        return Object.assign(Object.assign({}, super.getDefaultConfig()), { skipTests: true, pathToFile: '../db/db.json' });
    }
    get pathToFile() {
        if (path.isAbsolute(this.config.pathToFile)) {
            return this.config.pathToFile;
        }
        // Make sure the pathToFile is applied relative to the dist-dir
        return path.join(process.cwd(), 'dist', this.config.pathToFile);
    }
    async initialize() {
        const pathToFileDir = path.dirname(this.pathToFile);
        const pathExists = async (pathToFile) => !!(await fs.promises.stat(pathToFile).catch(() => false));
        if (!(await pathExists(pathToFileDir))) {
            await fs.promises.mkdir(pathToFileDir, { recursive: true });
        }
        if (!(await pathExists(this.pathToFile))) {
            await fs.promises.writeFile(this.pathToFile, '[]');
        }
    }
    async getDbItem(primaryKey) {
        const fileDataStr = await fs.promises.readFile(this.pathToFile, 'utf8');
        const users = fileDataStr.length > 0 ? JSON.parse(fileDataStr) : [];
        return users.find((userItem) => {
            return userItem.id === primaryKey;
        });
    }
    async loadData(userId, jovo) {
        const dbItem = await this.getDbItem(userId);
        if (dbItem) {
            jovo.$user.isNew = false;
            jovo.setPersistableData(dbItem, this.config.storedElements);
        }
    }
    async saveData(userId, jovo) {
        const fileDataStr = await fs.promises.readFile(this.pathToFile, 'utf8');
        const users = fileDataStr.length > 0 ? JSON.parse(fileDataStr) : [];
        const dbItem = users.find((userItem) => {
            return userItem.id === userId;
        });
        // // create new user
        if (!dbItem) {
            const item = {
                id: userId,
            };
            await this.applyPersistableData(jovo, item);
            users.push(item);
        }
        else {
            // update existing user
            for (let i = 0; i < users.length; i++) {
                if (users[i].id === userId) {
                    await this.applyPersistableData(jovo, users[i]);
                }
            }
        }
        return fs.promises.writeFile(this.pathToFile, JSON.stringify(users, null, 2));
    }
}
