// @jovotech/output/dist/esm2015/decorators/validation/IsSomeValid.js
import { formatList, isDefined, registerDecorator, } from '../..';
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function IsSomeValid(options, validationOptions) {
    if (options.keys.length <= 1) {
        throw new Error('At least 2 keys have to be defined in order to use IsSomeValid.');
    }
    return function (object, propertyKey) {
        registerDecorator({
            name: options.name || 'isSomeValid',
            target: object.constructor,
            propertyName: propertyKey.toString(),
            constraints: [options],
            options: validationOptions,
            async: true,
            validator: {
                async validate(value, args) {
                    const { keys, validate } = args.constraints[0];
                    const otherKeys = keys.filter((key) => {
                        return key !== args.property;
                    });
                    const otherPropertyMap = {};
                    for (const key of otherKeys) {
                        otherPropertyMap[key] = args.object[key];
                    }
                    const otherDefinedPropertyPairs = Object.entries(otherPropertyMap).filter((entry) => {
                        return isDefined(entry[1]);
                    });
                    // check if none are defined
                    if (!isDefined(value) && !otherDefinedPropertyPairs.length) {
                        const keysText = formatList(keys);
                        args.constraints[1] = `At least one of the properties ${keysText} must be defined. None is set.`;
                        return false;
                    }
                    // return true to skip validation for this property and not throw an error.
                    if (!isDefined(value)) {
                        return true;
                    }
                    const validationResult = await validate(value, args);
                    if (validationResult) {
                        args.constraints[1] = validationResult;
                        return false;
                    }
                    return true;
                },
                defaultMessage(args) {
                    return args.constraints[1];
                },
            },
        });
    };
}
