// @jovotech/platform-alexa/dist/esm2015/AlexaDevice.js
import { JovoDevice } from '@jovotech/framework';
import { getDeviceLocation, getDeviceAddress, } from './api/DeviceLocationApi';
import { getSystemTimeZone } from './api/SettingsApi';
export var AlexaCapability;
(function (AlexaCapability) {
    AlexaCapability["Apl"] = "ALEXA:APL";
    AlexaCapability["Html"] = "ALEXA:HTML";
})(AlexaCapability || (AlexaCapability = {}));
export class AlexaDevice extends JovoDevice {
    get id() {
        var _a;
        return (_a = this.jovo.$request.context) === null || _a === void 0 ? void 0 : _a.System.device.deviceId;
    }
    async getLocation() {
        const request = this.jovo.$request;
        return getDeviceLocation(request.getApiEndpoint(), request.getDeviceId(), request.getApiAccessToken());
    }
    async getAddress() {
        const request = this.jovo.$request;
        return getDeviceAddress(request.getApiEndpoint(), request.getDeviceId(), request.getApiAccessToken());
    }
    async getTimeZone() {
        const request = this.jovo.$request;
        return getSystemTimeZone(request.getApiEndpoint(), request.getDeviceId(), request.getApiAccessToken());
    }
}
