// @jovotech/platform-alexa/dist/esm2015/AlexaIsp.js
import { getProductList, } from './api/IspApi';
export class AlexaIsp {
    constructor(alexa) {
        this.alexa = alexa;
    }
    async getProductList(params) {
        return getProductList(this.alexa.$request.getApiEndpoint(), this.alexa.$request.getApiAccessToken(), this.alexa.$request.getLocale(), params);
    }
    async getProductByReferenceName(referenceName) {
        const products = await this.getProductList();
        return products.inSkillProducts.find((product) => product.referenceName === referenceName);
    }
    getPurchaseResult() {
        var _a, _b;
        return (_b = (_a = this.alexa.$request.request) === null || _a === void 0 ? void 0 : _a.payload) === null || _b === void 0 ? void 0 : _b.purchaseResult;
    }
    getProductId() {
        var _a, _b;
        return (_b = (_a = this.alexa.$request.request) === null || _a === void 0 ? void 0 : _a.payload) === null || _b === void 0 ? void 0 : _b.productId;
    }
    toJSON() {
        return Object.assign(Object.assign({}, this), { alexa: undefined });
    }
}
