// @jovotech/platform-googleassistant/dist/esm2015/GoogleAssistantPlatform.js
import { Platform } from '@jovotech/framework';
import _mergeWith from 'lodash.mergewith';
import { v4 as uuidV4 } from 'uuid';
import { GoogleAssistant } from './GoogleAssistant';
import { GoogleAssistantDevice } from './GoogleAssistantDevice';
import { GoogleAssistantRepromptComponent } from './GoogleAssistantRepromptComponent';
import { GoogleAssistantRequest } from './GoogleAssistantRequest';
import { GoogleAssistantRequestBuilder } from './GoogleAssistantRequestBuilder';
import { GoogleAssistantUser } from './GoogleAssistantUser';
import { GoogleAssistantOutputTemplateConverterStrategy } from './output';
export class GoogleAssistantPlatform extends Platform {
    constructor() {
        super(...arguments);
        this.id = 'googleAssistant';
        this.outputTemplateConverterStrategy = new GoogleAssistantOutputTemplateConverterStrategy();
        this.requestClass = GoogleAssistantRequest;
        this.jovoClass = GoogleAssistant;
        this.userClass = GoogleAssistantUser;
        this.deviceClass = GoogleAssistantDevice;
        this.requestBuilder = GoogleAssistantRequestBuilder;
    }
    getDefaultConfig() {
        return {};
    }
    mount(parent) {
        super.mount(parent);
        parent.middlewareCollection.use('before.request.start', (jovo) => {
            var _a, _b;
            if (((_b = (_a = jovo.$googleAssistant) === null || _a === void 0 ? void 0 : _a.$request.intent) === null || _b === void 0 ? void 0 : _b.name) === 'actions.intent.HEALTH_CHECK') {
                jovo.$handleRequest.stopMiddlewareExecution();
                return jovo.$handleRequest.server.setResponse({
                    prompt: { override: true, firstSimple: { speech: 'ok', text: '' } },
                });
            }
        });
        this.middlewareCollection.use('request.start', (jovo) => {
            return this.onRequestStart(jovo);
        });
    }
    initialize(parent) {
        parent.use(GoogleAssistantRepromptComponent);
    }
    isRequestRelated(request) {
        return request.user && request.session && request.handler && request.device;
    }
    isResponseRelated(response) {
        return response.user && response.session && response.prompt;
    }
    finalizeResponse(response, googleAssistant) {
        var _a;
        const requestSession = googleAssistant.$request.session || {};
        const responseSession = response.session || {};
        response.session = _mergeWith(Object.assign({ id: '', languageCode: '' }, requestSession), responseSession, { params: Object.assign({}, googleAssistant.$session) }, (objValue, srcValue) => {
            if (typeof objValue === 'string' && typeof srcValue === 'string') {
                return objValue ? objValue : srcValue;
            }
            // Replaces full array with the value furthest to the right. No concatenation of array values.
            if (Array.isArray(objValue) && Array.isArray(srcValue)) {
                return srcValue;
            }
        });
        if (googleAssistant.$request.user) {
            response.user = Object.assign({}, googleAssistant.$request.user);
        }
        if (response.scene && ((_a = googleAssistant.$request.scene) === null || _a === void 0 ? void 0 : _a.name)) {
            response.scene.name = googleAssistant.$request.scene.name;
        }
        return response;
    }
    onRequestStart(jovo) {
        var _a, _b, _c, _d, _e, _f, _g;
        const user = (_a = jovo.$googleAssistant) === null || _a === void 0 ? void 0 : _a.$user;
        // if the user is linked and has no user id, generate one
        if (user && user.isVerified() && !user.id) {
            user.setId(uuidV4());
        }
        const request = (_b = jovo.$googleAssistant) === null || _b === void 0 ? void 0 : _b.$request;
        // if it is a selection-event
        if ((request === null || request === void 0 ? void 0 : request.intent) &&
            !((_c = request === null || request === void 0 ? void 0 : request.intent) === null || _c === void 0 ? void 0 : _c.name) &&
            !!((_d = request.scene) === null || _d === void 0 ? void 0 : _d.slotFillingStatus) &&
            Object.keys(((_e = request.intent) === null || _e === void 0 ? void 0 : _e.params) || {}).length &&
            ((_g = (_f = request.session) === null || _f === void 0 ? void 0 : _f.params) === null || _g === void 0 ? void 0 : _g._GOOGLE_ASSISTANT_SELECTION_INTENT_)) {
            jovo.$input.intent = request.session.params._GOOGLE_ASSISTANT_SELECTION_INTENT_;
            delete request.session.params._GOOGLE_ASSISTANT_SELECTION_INTENT_;
        }
    }
}
