// @jovotech/framework/dist/esm5/plugins/DbPlugin.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { HandleRequest, InvalidParentError } from '..';
import { Plugin } from '../Plugin';
export var DEFAULT_SESSION_EXPIRES_AFTER_SECONDS = 900;
var DbPlugin = /** @class */ (function (_super) {
    __extends(DbPlugin, _super);
    function DbPlugin(config) {
        var _this = _super.call(this, config) || this;
        for (var prop in _this.config.storedElements) {
            if (_this.config.storedElements.hasOwnProperty(prop)) {
                if (_this.config.storedElements[prop] &&
                    typeof _this.config.storedElements[prop] === 'boolean') {
                    _this.config.storedElements[prop] = {
                        enabled: _this.config.storedElements[prop],
                    };
                }
                if (_this.config.storedElements[prop] &&
                    typeof _this.config.storedElements[prop] === 'object' &&
                    typeof _this.config.storedElements[prop].enabled === 'undefined') {
                    _this.config.storedElements[prop].enabled = true;
                }
            }
        }
        return _this;
    }
    DbPlugin.prototype.getDefaultConfig = function () {
        return {
            enabled: true,
            storedElements: {
                user: {
                    enabled: true,
                },
                session: {
                    enabled: false,
                    expiresAfterSeconds: DEFAULT_SESSION_EXPIRES_AFTER_SECONDS,
                },
                history: {
                    size: 3,
                    enabled: false,
                    output: true,
                    nlu: true,
                    asr: false,
                    state: false,
                    request: false,
                    response: false,
                },
                createdAt: true,
                updatedAt: true,
            },
        };
    };
    DbPlugin.prototype.mount = function (parent) {
        var _this = this;
        if (!(parent instanceof HandleRequest)) {
            throw new InvalidParentError(this.name, HandleRequest);
        }
        parent.middlewareCollection.use('request.end', function (jovo) {
            if (!jovo.$user.id) {
                return;
            }
            return _this.loadData(jovo.$user.id, jovo);
        });
        parent.middlewareCollection.use('response.end', function (jovo) {
            if (!jovo.$user.id) {
                return;
            }
            return _this.saveData(jovo.$user.id, jovo);
        });
    };
    DbPlugin.prototype.applyPersistableData = function (jovo, item) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
        return __awaiter(this, void 0, void 0, function () {
            var persistableData, _l, _m, _i, prop, historyLastItem, newHistoryItem, _o, _p, _q, propHistory, reservedPropertyNames, func, _r, _s;
            return __generator(this, function (_t) {
                switch (_t.label) {
                    case 0:
                        persistableData = jovo.getPersistableData();
                        _l = [];
                        for (_m in persistableData)
                            _l.push(_m);
                        _i = 0;
                        _t.label = 1;
                    case 1:
                        if (!(_i < _l.length)) return [3 /*break*/, 9];
                        prop = _l[_i];
                        if (!(persistableData.hasOwnProperty(prop) &&
                            (((_a = this.config.storedElements) === null || _a === void 0 ? void 0 : _a[prop]) === true ||
                                ((_b = this.config.storedElements) === null || _b === void 0 ? void 0 : _b[prop]).enabled))) return [3 /*break*/, 8];
                        if (!(prop === 'history')) return [3 /*break*/, 7];
                        historyLastItem = jovo.getCurrentHistoryItem();
                        newHistoryItem = {};
                        if (!(((_c = this.config.storedElements) === null || _c === void 0 ? void 0 : _c.history) &&
                            typeof ((_d = this.config.storedElements) === null || _d === void 0 ? void 0 : _d.history) === 'object')) return [3 /*break*/, 6];
                        _o = [];
                        for (_p in (_e = this.config.storedElements) === null || _e === void 0 ? void 0 : _e.history)
                            _o.push(_p);
                        _q = 0;
                        _t.label = 2;
                    case 2:
                        if (!(_q < _o.length)) return [3 /*break*/, 6];
                        propHistory = _o[_q];
                        reservedPropertyNames = ['size', 'enabled'];
                        if (!((_f = this.config.storedElements) === null || _f === void 0 ? void 0 : _f.history.hasOwnProperty(propHistory))) return [3 /*break*/, 5];
                        if (reservedPropertyNames.includes(propHistory)) {
                            return [3 /*break*/, 5];
                        }
                        if (!((_g = this.config.storedElements) === null || _g === void 0 ? void 0 : _g.history[propHistory])) return [3 /*break*/, 5];
                        if (!(typeof ((_h = this.config.storedElements) === null || _h === void 0 ? void 0 : _h.history[propHistory]) === 'function')) return [3 /*break*/, 4];
                        func = (_j = this.config.storedElements) === null || _j === void 0 ? void 0 : _j.history[propHistory];
                        _r = newHistoryItem;
                        _s = propHistory;
                        return [4 /*yield*/, func(jovo)];
                    case 3:
                        _r[_s] = _t.sent();
                        return [3 /*break*/, 5];
                    case 4:
                        // default history item like nlu, output etc
                        newHistoryItem[propHistory] = historyLastItem[propHistory];
                        _t.label = 5;
                    case 5:
                        _q++;
                        return [3 /*break*/, 2];
                    case 6:
                        // put latest history item on first position in array
                        item[prop] = {
                            items: [newHistoryItem].concat(persistableData.history.items),
                        };
                        // remove trailing history items
                        item[prop].items = item[prop].items.slice(0, ((_k = this.config.storedElements) === null || _k === void 0 ? void 0 : _k.history).size);
                        return [3 /*break*/, 8];
                    case 7:
                        item[prop] = persistableData[prop];
                        _t.label = 8;
                    case 8:
                        _i++;
                        return [3 /*break*/, 1];
                    case 9: return [2 /*return*/];
                }
            });
        });
    };
    return DbPlugin;
}(Plugin));
export { DbPlugin };
