// @jovotech/platform-alexa/dist/esm2015/output/decorators/validation/IsValidDynamicEntitiesSlotTypesArray.js
import { formatValidationErrors, isDefined, registerDecorator, validate, } from '@jovotech/output';
import { DynamicEntitiesUpdateBehavior, } from '../../models/dialog/DialogUpdateDynamicEntitiesDirective';
export function IsValidDynamicEntitiesSlotTypesArray(options) {
    return function (object, propertyKey) {
        registerDecorator({
            name: 'isValidDynamicEntitiesSlotTypesArray',
            target: object.constructor,
            propertyName: propertyKey.toString(),
            constraints: [],
            options,
            async: true,
            validator: {
                async validate(value, args) {
                    const behavior = args.object.updateBehavior;
                    if (behavior === DynamicEntitiesUpdateBehavior.Replace) {
                        if (!Array.isArray(value)) {
                            args.constraints[0] = '$property must be an array';
                            return false;
                        }
                        const errors = await validate(value);
                        args.constraints[0] = formatValidationErrors(errors, {
                            text: '$property is invalid:',
                            delimiter: '\n  - ',
                            path: '$property',
                        });
                        return !errors.length;
                    }
                    else {
                        if (isDefined(value)) {
                            args.constraints[0] = `$property can not be set when updateBehavior is set to ${behavior}`;
                            return false;
                        }
                    }
                    return true;
                },
                defaultMessage(args) {
                    return args.constraints[0];
                },
            },
        });
    };
}
