// @jovotech/platform-alexa/dist/esm2015/output/decorators/validation/IsValidCardString.js
import { isDefined, registerDecorator, } from '@jovotech/output';
import { validateAlexaString } from '../../utilities';
export function IsValidCardString(relatedTypes, options) {
    return function (object, propertyKey) {
        registerDecorator({
            name: 'isValidCardString',
            target: object.constructor,
            propertyName: propertyKey.toString(),
            constraints: [],
            options,
            validator: {
                validate(value, args) {
                    const type = args.object.type;
                    // if there is no type, skip for now because another decorator should take care of that
                    if (!type) {
                        return true;
                    }
                    if (isDefined(value) && !relatedTypes.includes(type)) {
                        args.constraints[0] = `$property can not be set when the type is ${type}`;
                        return false;
                    }
                    if (isDefined(value)) {
                        const result = validateAlexaString(value);
                        if (result) {
                            args.constraints[0] = result;
                            return false;
                        }
                    }
                    return true;
                },
                defaultMessage(args) {
                    return args.constraints[0];
                },
            },
        });
    };
}
