// @jovotech/platform-alexa/dist/esm2015/output/decorators/validation/MainTextMaxLength.js
import { registerDecorator } from '@jovotech/output';
import { TextContent } from '../../models';
export function MainTextMaxLength(length, options) {
    return function (object, propertyKey) {
        registerDecorator({
            name: 'mainTextMaxLength',
            target: object.constructor,
            propertyName: propertyKey.toString(),
            constraints: [length],
            options,
            validator: {
                validate(value, args) {
                    var _a, _b;
                    if (!(value instanceof TextContent)) {
                        return true;
                    }
                    if (((_b = (_a = value.primaryText) === null || _a === void 0 ? void 0 : _a.text) === null || _b === void 0 ? void 0 : _b.length) > args.constraints[0]) {
                        args.constraints[1] = `primaryText of $property can not exceed ${args.constraints[0]} characters`;
                        return false;
                    }
                    return true;
                },
                defaultMessage(args) {
                    return args.constraints[1];
                },
            },
        });
    };
}
