// @jovotech/platform-alexa/dist/esm5/AlexaUser.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { JovoUser } from '@jovotech/framework';
import { ProfileProperty, sendCustomerProfileApiRequest } from './api';
import { PersonProfileProperty, sendPersonProfileApiRequest, } from './api/PersonProfileApi';
import { deleteReminder, getAllReminders, getReminder, setReminder, updateReminder, } from './api/ReminderApi';
import { getListItem, getListItems, getLists } from './api/ListApi';
var AlexaUser = /** @class */ (function (_super) {
    __extends(AlexaUser, _super);
    function AlexaUser(jovo) {
        return _super.call(this, jovo) || this;
    }
    Object.defineProperty(AlexaUser.prototype, "id", {
        get: function () {
            return this.jovo.$request.getUserId();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AlexaUser.prototype, "accessToken", {
        get: function () {
            var _a;
            return (_a = this.jovo.$request.session) === null || _a === void 0 ? void 0 : _a.user.accessToken;
        },
        enumerable: false,
        configurable: true
    });
    AlexaUser.prototype.getEmail = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getProfileProperty(ProfileProperty.EMAIL)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    AlexaUser.prototype.getMobileNumber = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getProfileProperty(ProfileProperty.MOBILE_NUMBER)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    AlexaUser.prototype.getName = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getProfileProperty(ProfileProperty.NAME)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    AlexaUser.prototype.getGivenName = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getProfileProperty(ProfileProperty.GIVEN_NAME)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    AlexaUser.prototype.getSpeakerName = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getPersonProfileProperty(PersonProfileProperty.NAME)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    AlexaUser.prototype.getSpeakerGivenName = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getPersonProfileProperty(PersonProfileProperty.GIVEN_NAME)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    AlexaUser.prototype.getSpeakerMobileNumber = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getPersonProfileProperty(PersonProfileProperty.MOBILE_NUMBER)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    AlexaUser.prototype.getPersonProfileProperty = function (property) {
        return __awaiter(this, void 0, void 0, function () {
            var request;
            return __generator(this, function (_a) {
                request = this.jovo.$request;
                return [2 /*return*/, sendPersonProfileApiRequest(property, request.getApiEndpoint(), request.getApiAccessToken())];
            });
        });
    };
    AlexaUser.prototype.getProfileProperty = function (property) {
        return __awaiter(this, void 0, void 0, function () {
            var request;
            return __generator(this, function (_a) {
                request = this.jovo.$request;
                return [2 /*return*/, sendCustomerProfileApiRequest(property, request.getApiEndpoint(), request.getApiAccessToken())];
            });
        });
    };
    AlexaUser.prototype.setReminder = function (reminder) {
        return __awaiter(this, void 0, void 0, function () {
            var request;
            return __generator(this, function (_a) {
                request = this.jovo.$request;
                return [2 /*return*/, setReminder(reminder, request.getApiEndpoint(), request.getApiAccessToken())];
            });
        });
    };
    AlexaUser.prototype.updateReminder = function (alertToken, reminder) {
        return __awaiter(this, void 0, void 0, function () {
            var request;
            return __generator(this, function (_a) {
                request = this.jovo.$request;
                return [2 /*return*/, updateReminder(alertToken, reminder, request.getApiEndpoint(), request.getApiAccessToken())];
            });
        });
    };
    AlexaUser.prototype.deleteReminder = function (alertToken) {
        return __awaiter(this, void 0, void 0, function () {
            var request;
            return __generator(this, function (_a) {
                request = this.jovo.$request;
                return [2 /*return*/, deleteReminder(alertToken, request.getApiEndpoint(), request.getApiAccessToken())];
            });
        });
    };
    AlexaUser.prototype.getAllReminders = function () {
        return __awaiter(this, void 0, void 0, function () {
            var request;
            return __generator(this, function (_a) {
                request = this.jovo.$request;
                return [2 /*return*/, getAllReminders(request.getApiEndpoint(), request.getApiAccessToken())];
            });
        });
    };
    AlexaUser.prototype.getReminder = function (alertToken) {
        return __awaiter(this, void 0, void 0, function () {
            var request;
            return __generator(this, function (_a) {
                request = this.jovo.$request;
                return [2 /*return*/, getReminder(alertToken, request.getApiEndpoint(), request.getApiAccessToken())];
            });
        });
    };
    AlexaUser.prototype.getLists = function () {
        var request = this.jovo.$request;
        return getLists(request.getApiEndpoint(), request.getApiAccessToken());
    };
    AlexaUser.prototype.getListItem = function (listId, itemId) {
        var request = this.jovo.$request;
        return getListItem(listId, itemId, request.getApiEndpoint(), request.getApiAccessToken());
    };
    AlexaUser.prototype.getListItems = function (listId, itemIds) {
        var request = this.jovo.$request;
        return getListItems(listId, itemIds, request.getApiEndpoint(), request.getApiAccessToken());
    };
    return AlexaUser;
}(JovoUser));
export { AlexaUser };
