// @jovotech/framework/dist/esm5/errors/MatchingRouteNotFoundError.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { JovoError } from '@jovotech/common';
var MatchingRouteNotFoundError = /** @class */ (function (_super) {
    __extends(MatchingRouteNotFoundError, _super);
    function MatchingRouteNotFoundError(_a) {
        var request = _a.request, input = _a.input, state = _a.state, matches = _a.matches;
        return _super.call(this, {
            message: 'No matching route was found for the request.',
            context: {
                input: input,
                state: state,
                matches: matches,
                request: request,
            },
        }) || this;
    }
    return MatchingRouteNotFoundError;
}(JovoError));
export { MatchingRouteNotFoundError };
