// @jovotech/framework/dist/esm5/testsuite/TestRequest.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { InputType, } from '@jovotech/common';
import { JovoRequest, JovoSession } from '..';
var TestRequest = /** @class */ (function (_super) {
    __extends(TestRequest, _super);
    function TestRequest() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.isTestRequest = true;
        _this.session = new JovoSession({ state: [] });
        return _this;
    }
    TestRequest.prototype.getLocale = function () {
        return this.locale;
    };
    TestRequest.prototype.setLocale = function (locale) {
        this.locale = locale;
    };
    TestRequest.prototype.getIntent = function () {
        return this.intent;
    };
    TestRequest.prototype.setIntent = function (intent) {
        this.intent = intent;
    };
    TestRequest.prototype.getEntities = function () {
        return;
    };
    TestRequest.prototype.getInputType = function () {
        return this.session.isNew ? InputType.Launch : InputType.Intent;
    };
    TestRequest.prototype.getInputText = function () {
        return;
    };
    TestRequest.prototype.getInputAudio = function () {
        return;
    };
    TestRequest.prototype.setSessionData = function (data) {
        this.session.data = data;
    };
    TestRequest.prototype.getUserId = function () {
        return this.userId;
    };
    TestRequest.prototype.setUserId = function (userId) {
        this.userId = userId;
    };
    TestRequest.prototype.getSessionData = function () {
        return this.session.data;
    };
    TestRequest.prototype.getSessionId = function () {
        return this.session.id;
    };
    TestRequest.prototype.isNewSession = function () {
        return this.session.isNew;
    };
    TestRequest.prototype.getDeviceCapabilities = function () {
        return;
    };
    TestRequest.prototype.getRequestId = function () {
        return;
    };
    return TestRequest;
}(JovoRequest));
export { TestRequest };
