// @jovotech/platform-core/dist/esm2015/output/CoreOutputTemplateConverterStrategy.js
import { mergeInstances, mergeListen, OutputTemplateConverterStrategy, } from '@jovotech/output';
import { CoreResponse } from '../CoreResponse';
export class CoreOutputTemplateConverterStrategy extends OutputTemplateConverterStrategy {
    constructor() {
        super(...arguments);
        this.responseClass = CoreResponse;
        this.platformName = 'core';
    }
    toResponse(output) {
        output = Array.isArray(output) ? output : [output];
        const response = this.normalizeResponse({
            version: '4.0.0',
            platform: 'core',
            output,
            context: {
                request: {},
                session: {
                    end: false,
                    data: {},
                },
                user: {
                    data: {},
                },
            },
        });
        let mergedListen;
        output.forEach((outputItem) => {
            var _a, _b;
            mergedListen = mergeListen(mergedListen, outputItem.listen);
            if ((_b = (_a = outputItem.platforms) === null || _a === void 0 ? void 0 : _a.core) === null || _b === void 0 ? void 0 : _b.nativeResponse) {
                mergeInstances(response, outputItem.platforms.core.nativeResponse);
            }
        });
        response.context.session.end = !(mergedListen !== null && mergedListen !== void 0 ? mergedListen : true);
        return response;
    }
    fromResponse(response) {
        return response.output;
    }
}
