// @jovotech/platform-alexa/dist/cjs/AlexaTask.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlexaTask = void 0;
class AlexaTask {
    constructor(alexa) {
        this.alexa = alexa;
    }
    getTask() {
        var _a;
        return (_a = this.alexa.$request.request) === null || _a === void 0 ? void 0 : _a.task;
    }
    get version() {
        var _a;
        return (_a = this.getTask()) === null || _a === void 0 ? void 0 : _a.version;
    }
    get name() {
        var _a;
        return (_a = this.getTask()) === null || _a === void 0 ? void 0 : _a.name;
    }
    get input() {
        var _a;
        return (_a = this.getTask()) === null || _a === void 0 ? void 0 : _a.input;
    }
    /**
     * Looks at the task name and checks if it ends with the provided taskName. Ignores skillId prefix
     * @param taskName Task name without skill id
     * @returns true, if task name comes after the skill id
     */
    hasTaskName(taskName) {
        if (!this.name)
            return false;
        return this.name.endsWith(`.${taskName}`);
    }
    isVersion(version) {
        if (!this.version)
            return false;
        return this.version === `${version}`;
    }
    toJSON() {
        return Object.assign(Object.assign({}, this), { alexa: undefined });
    }
}
exports.AlexaTask = AlexaTask;
