// @jovotech/output/dist/esm5/models/Message.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { IsSomeValid, isString } from '..';
export var IsValidMessageString = function () {
    return IsSomeValid({
        keys: ['speech', 'text'],
        validate: function (value) {
            if (!isString(value)) {
                return '$property must be a string';
            }
            if (!value) {
                return '$property should not be empty';
            }
            return;
        },
    });
};
var Message = /** @class */ (function () {
    function Message() {
    }
    __decorate([
        IsValidMessageString(),
        __metadata("design:type", String)
    ], Message.prototype, "speech", void 0);
    __decorate([
        IsValidMessageString(),
        __metadata("design:type", String)
    ], Message.prototype, "text", void 0);
    return Message;
}());
export { Message };
var SpeechMessage = /** @class */ (function (_super) {
    __extends(SpeechMessage, _super);
    function SpeechMessage() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return SpeechMessage;
}(Message));
export { SpeechMessage };
var TextMessage = /** @class */ (function (_super) {
    __extends(TextMessage, _super);
    function TextMessage() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return TextMessage;
}(Message));
export { TextMessage };
