// @jovotech/platform-googleassistant/dist/esm2015/cli/hooks/GetHook.js
import { ANSWER_CANCEL, execAsync, flags, JovoCliError, PluginHook, printHighlight, promptOverwrite, Task, } from '@jovotech/cli-core';
import { existsSync, mkdirSync } from 'fs';
import _get from 'lodash.get';
import { checkForGactionsCli, getGactionsError } from '../utilities';
export class GetHook extends PluginHook {
    install() {
        this.middlewareCollection = {
            'install': [this.addCliOptions.bind(this)],
            'before.get:platform': [
                this.checkForPlatform.bind(this),
                checkForGactionsCli,
                this.updatePluginContext.bind(this),
                this.checkForExistingPlatformFiles.bind(this),
            ],
            'get:platform': [this.get.bind(this)],
        };
    }
    /**
     * Add platform-specific CLI options, including flags and args.
     * @param context - Context providing an access point to command flags and args.
     */
    addCliOptions(context) {
        if (context.command !== 'get:platform') {
            return;
        }
        context.flags['project-id'] = flags.string({
            description: 'Google Cloud Project ID',
        });
    }
    /**
     * Checks if the currently selected platform matches this CLI plugin.
     * @param context - Context containing information after flags and args have been parsed by the CLI.
     */
    checkForPlatform() {
        // Check if this plugin should be used or not.
        if (!this.$context.platforms.includes(this.$plugin.id)) {
            this.uninstall();
        }
    }
    /**
     * Updates the current plugin context with platform-specific values.
     */
    updatePluginContext() {
        if (!this.$context.googleAssistant) {
            this.$context.googleAssistant = {};
        }
        this.$context.googleAssistant.projectId =
            this.$context.flags['project-id'] || _get(this.$plugin.config, 'projectId');
        if (!this.$context.googleAssistant.projectId) {
            throw new JovoCliError({
                message: 'Could not find projectId.',
                module: 'GoogleAssistantCli',
                hint: 'Please provide a project id by using the flag "--project-id" or in your project configuration.',
            });
        }
    }
    /**
     * Checks if platform-specific files already exist and prompts for overwriting them.
     */
    async checkForExistingPlatformFiles() {
        if (!this.$context.flags.clean && existsSync(this.$plugin.platformPath)) {
            const answer = await promptOverwrite('Found existing GoogleAssistant project files. How to proceed?');
            if (answer.overwrite === ANSWER_CANCEL) {
                this.uninstall();
            }
        }
    }
    /**
     * Fetches platform-specific models, such as intents and entities from the Google Actions Console.
     */
    async get() {
        const getTask = new Task(`Getting Conversational Actions Project ${printHighlight(`(${this.$context.googleAssistant.projectId})`)}`, async () => {
            const platformPath = this.$plugin.platformPath;
            if (!existsSync(platformPath)) {
                mkdirSync(platformPath);
            }
            try {
                await execAsync(`gactions pull --clean --force --project-id ${this.$context.googleAssistant.projectId} --consumer jovo-cli`, { cwd: platformPath });
            }
            catch (error) {
                throw getGactionsError(error.stderr);
            }
        });
        await getTask.run();
    }
}
