// @jovotech/platform-googleassistant/dist/esm5/GoogleAssistantRequest.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Capability, InputType, JovoRequest, } from '@jovotech/framework';
import { GoogleAssistantSystemInputType, GoogleAssistantSystemIntent } from './enums';
import { GoogleAssistantCapability } from './GoogleAssistantDevice';
import { Capability as NativeCapability } from './output';
var GoogleAssistantRequest = /** @class */ (function (_super) {
    __extends(GoogleAssistantRequest, _super);
    function GoogleAssistantRequest() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    GoogleAssistantRequest.prototype.getLocale = function () {
        var _a;
        return (_a = this.user) === null || _a === void 0 ? void 0 : _a.locale;
    };
    GoogleAssistantRequest.prototype.setLocale = function (locale) {
        if (!this.user) {
            return;
        }
        this.user.locale = locale;
    };
    GoogleAssistantRequest.prototype.getIntent = function () {
        var _a;
        return (_a = this.intent) === null || _a === void 0 ? void 0 : _a.name;
    };
    GoogleAssistantRequest.prototype.setIntent = function (intent) {
        if (!this.intent) {
            this.intent = { name: intent, params: {} };
        }
        this.intent.name = intent;
    };
    GoogleAssistantRequest.prototype.getEntities = function () {
        var _a, _b;
        var entities = {};
        for (var param in (_a = this.intent) === null || _a === void 0 ? void 0 : _a.params) {
            if ((_b = this.intent) === null || _b === void 0 ? void 0 : _b.params.hasOwnProperty(param)) {
                entities[param] = {
                    native: this.intent.params[param],
                    id: this.intent.params[param].resolved,
                    value: this.intent.params[param].original,
                    resolved: this.intent.params[param].resolved,
                };
            }
        }
        return entities;
    };
    GoogleAssistantRequest.prototype.getInputType = function () {
        var _a, _b, _c, _d;
        if (((_a = this.intent) === null || _a === void 0 ? void 0 : _a.name) === GoogleAssistantSystemIntent.Main &&
            !Object.keys(((_b = this.session) === null || _b === void 0 ? void 0 : _b.params) || {}).length) {
            return InputType.Launch;
        }
        if (((_c = this.intent) === null || _c === void 0 ? void 0 : _c.name) === GoogleAssistantSystemIntent.Cancel) {
            return InputType.End;
        }
        if ((_d = this.intent) === null || _d === void 0 ? void 0 : _d.params.AccountLinkingSlot) {
            return GoogleAssistantSystemInputType.ON_SIGN_IN;
        }
        return undefined;
    };
    GoogleAssistantRequest.prototype.getInputText = function () {
        var _a;
        return (_a = this.intent) === null || _a === void 0 ? void 0 : _a.query;
    };
    GoogleAssistantRequest.prototype.getInputAudio = function () {
        return;
    };
    GoogleAssistantRequest.prototype.getSessionData = function () {
        var _a;
        return (_a = this.session) === null || _a === void 0 ? void 0 : _a.params;
    };
    GoogleAssistantRequest.prototype.setSessionData = function (data) {
        if (!this.session) {
            return;
        }
        this.session.params = data;
    };
    GoogleAssistantRequest.prototype.getSessionId = function () {
        var _a;
        return (_a = this.session) === null || _a === void 0 ? void 0 : _a.id;
    };
    GoogleAssistantRequest.prototype.isNewSession = function () {
        var sessionData = this.getSessionData();
        return !sessionData || Object.keys(sessionData).length === 0;
    };
    GoogleAssistantRequest.prototype.getDeviceCapabilities = function () {
        var _a;
        var supportedCapabilities = (_a = this.device) === null || _a === void 0 ? void 0 : _a.capabilities;
        if (!supportedCapabilities) {
            return;
        }
        var capabilities = [];
        if (supportedCapabilities === null || supportedCapabilities === void 0 ? void 0 : supportedCapabilities.includes(NativeCapability.Speech)) {
            capabilities.push(Capability.Audio);
        }
        if (supportedCapabilities === null || supportedCapabilities === void 0 ? void 0 : supportedCapabilities.includes(NativeCapability.LongFormAudio)) {
            capabilities.push(Capability.LongformAudio);
        }
        if (supportedCapabilities === null || supportedCapabilities === void 0 ? void 0 : supportedCapabilities.includes(NativeCapability.RichResponse)) {
            capabilities.push(Capability.Screen);
        }
        if (supportedCapabilities === null || supportedCapabilities === void 0 ? void 0 : supportedCapabilities.includes(NativeCapability.WebLink)) {
            capabilities.push(GoogleAssistantCapability.WebLink);
        }
        if (supportedCapabilities === null || supportedCapabilities === void 0 ? void 0 : supportedCapabilities.includes(NativeCapability.InteractiveCanvas)) {
            capabilities.push(GoogleAssistantCapability.InteractiveCanvas);
        }
        return capabilities;
    };
    GoogleAssistantRequest.prototype.getUserId = function () {
        var _a, _b;
        return (_b = (_a = this.user) === null || _a === void 0 ? void 0 : _a.params) === null || _b === void 0 ? void 0 : _b.userId;
    };
    GoogleAssistantRequest.prototype.setUserId = function (userId) {
        var _a;
        if (!((_a = this.user) === null || _a === void 0 ? void 0 : _a.params)) {
            return;
        }
        this.user.params.userId = userId;
    };
    GoogleAssistantRequest.prototype.getRequestId = function () {
        return;
    };
    return GoogleAssistantRequest;
}(JovoRequest));
export { GoogleAssistantRequest };
