// @jovotech/platform-googleassistant/dist/esm5/GoogleAssistantPlatform.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { Platform } from '@jovotech/framework';
import _mergeWith from 'lodash.mergewith';
import { v4 as uuidV4 } from 'uuid';
import { GoogleAssistant } from './GoogleAssistant';
import { GoogleAssistantDevice } from './GoogleAssistantDevice';
import { GoogleAssistantRepromptComponent } from './GoogleAssistantRepromptComponent';
import { GoogleAssistantRequest } from './GoogleAssistantRequest';
import { GoogleAssistantRequestBuilder } from './GoogleAssistantRequestBuilder';
import { GoogleAssistantUser } from './GoogleAssistantUser';
import { GoogleAssistantOutputTemplateConverterStrategy } from './output';
var GoogleAssistantPlatform = /** @class */ (function (_super) {
    __extends(GoogleAssistantPlatform, _super);
    function GoogleAssistantPlatform() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.id = 'googleAssistant';
        _this.outputTemplateConverterStrategy = new GoogleAssistantOutputTemplateConverterStrategy();
        _this.requestClass = GoogleAssistantRequest;
        _this.jovoClass = GoogleAssistant;
        _this.userClass = GoogleAssistantUser;
        _this.deviceClass = GoogleAssistantDevice;
        _this.requestBuilder = GoogleAssistantRequestBuilder;
        return _this;
    }
    GoogleAssistantPlatform.prototype.getDefaultConfig = function () {
        return {};
    };
    GoogleAssistantPlatform.prototype.mount = function (parent) {
        var _this = this;
        _super.prototype.mount.call(this, parent);
        parent.middlewareCollection.use('before.request.start', function (jovo) {
            var _a, _b;
            if (((_b = (_a = jovo.$googleAssistant) === null || _a === void 0 ? void 0 : _a.$request.intent) === null || _b === void 0 ? void 0 : _b.name) === 'actions.intent.HEALTH_CHECK') {
                jovo.$handleRequest.stopMiddlewareExecution();
                return jovo.$handleRequest.server.setResponse({
                    prompt: { override: true, firstSimple: { speech: 'ok', text: '' } },
                });
            }
        });
        this.middlewareCollection.use('request.start', function (jovo) {
            return _this.onRequestStart(jovo);
        });
    };
    GoogleAssistantPlatform.prototype.initialize = function (parent) {
        parent.use(GoogleAssistantRepromptComponent);
    };
    GoogleAssistantPlatform.prototype.isRequestRelated = function (request) {
        return request.user && request.session && request.handler && request.device;
    };
    GoogleAssistantPlatform.prototype.isResponseRelated = function (response) {
        return response.user && response.session && response.prompt;
    };
    GoogleAssistantPlatform.prototype.finalizeResponse = function (response, googleAssistant) {
        var _a;
        var requestSession = googleAssistant.$request.session || {};
        var responseSession = response.session || {};
        response.session = _mergeWith(__assign({ id: '', languageCode: '' }, requestSession), responseSession, { params: __assign({}, googleAssistant.$session) }, function (objValue, srcValue) {
            if (typeof objValue === 'string' && typeof srcValue === 'string') {
                return objValue ? objValue : srcValue;
            }
            // Replaces full array with the value furthest to the right. No concatenation of array values.
            if (Array.isArray(objValue) && Array.isArray(srcValue)) {
                return srcValue;
            }
        });
        if (googleAssistant.$request.user) {
            response.user = __assign({}, googleAssistant.$request.user);
        }
        if (response.scene && ((_a = googleAssistant.$request.scene) === null || _a === void 0 ? void 0 : _a.name)) {
            response.scene.name = googleAssistant.$request.scene.name;
        }
        return response;
    };
    GoogleAssistantPlatform.prototype.onRequestStart = function (jovo) {
        var _a, _b, _c, _d, _e, _f, _g;
        var user = (_a = jovo.$googleAssistant) === null || _a === void 0 ? void 0 : _a.$user;
        // if the user is linked and has no user id, generate one
        if (user && user.isVerified() && !user.id) {
            user.setId(uuidV4());
        }
        var request = (_b = jovo.$googleAssistant) === null || _b === void 0 ? void 0 : _b.$request;
        // if it is a selection-event
        if ((request === null || request === void 0 ? void 0 : request.intent) &&
            !((_c = request === null || request === void 0 ? void 0 : request.intent) === null || _c === void 0 ? void 0 : _c.name) &&
            !!((_d = request.scene) === null || _d === void 0 ? void 0 : _d.slotFillingStatus) &&
            Object.keys(((_e = request.intent) === null || _e === void 0 ? void 0 : _e.params) || {}).length &&
            ((_g = (_f = request.session) === null || _f === void 0 ? void 0 : _f.params) === null || _g === void 0 ? void 0 : _g._GOOGLE_ASSISTANT_SELECTION_INTENT_)) {
            jovo.$input.intent = request.session.params._GOOGLE_ASSISTANT_SELECTION_INTENT_;
            delete request.session.params._GOOGLE_ASSISTANT_SELECTION_INTENT_;
        }
    };
    return GoogleAssistantPlatform;
}(Platform));
export { GoogleAssistantPlatform };
