// @jovotech/platform-alexa/dist/cjs/output/templates/AplRenderDocumentOutput.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AplRenderDocumentOutput = void 0;
const framework_1 = require("@jovotech/framework");
let AplRenderDocumentOutput = class AplRenderDocumentOutput extends framework_1.BaseOutput {
    async build() {
        const aplRenderDocDirective = {
            type: 'Alexa.Presentation.APL.RenderDocument',
            document: this.options.document,
            datasources: this.options.datasources,
            sources: this.options.sources,
            token: this.options.token || 'token',
        };
        return {
            platforms: {
                alexa: {
                    nativeResponse: {
                        response: {
                            directives: [aplRenderDocDirective],
                        },
                    },
                },
            },
        };
    }
};
AplRenderDocumentOutput = __decorate([
    (0, framework_1.Output)()
], AplRenderDocumentOutput);
exports.AplRenderDocumentOutput = AplRenderDocumentOutput;
