// @jovotech/platform-alexa/dist/cjs/output/templates/ConnectionAskForPermissionConsentOutput.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectionAskForPermissionConsentOutput = void 0;
const framework_1 = require("@jovotech/framework");
const OnCompletion_1 = require("../models/common/OnCompletion");
const StartConnectionOutput_1 = require("./StartConnectionOutput");
let ConnectionAskForPermissionConsentOutput = class ConnectionAskForPermissionConsentOutput extends framework_1.BaseOutput {
    getDefaultOptions() {
        return {
            onCompletion: OnCompletion_1.OnCompletion.ResumeSession,
            permissionScopes: [],
        };
    }
    build() {
        return new StartConnectionOutput_1.StartConnectionOutput(this.jovo, {
            taskName: { name: 'AskForPermissionsConsent', amazonPredefinedTask: true },
            taskVersion: 2,
            shouldEndSession: this.options.shouldEndSession,
            onCompletion: this.options.onCompletion,
            input: {
                '@type': 'AskForPermissionsConsentRequest',
                '@version': '2',
                'permissionScopes': this.options.permissionScopes,
            },
            token: this.options.token,
        }).build();
    }
};
ConnectionAskForPermissionConsentOutput = __decorate([
    (0, framework_1.Output)()
], ConnectionAskForPermissionConsentOutput);
exports.ConnectionAskForPermissionConsentOutput = ConnectionAskForPermissionConsentOutput;
