// @jovotech/platform-alexa/dist/cjs/output/templates/VideoAppLaunchOutput.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VideoAppLaunchOutput = void 0;
const framework_1 = require("@jovotech/framework");
let VideoAppLaunchOutput = class VideoAppLaunchOutput extends framework_1.BaseOutput {
    async build() {
        const videoAppLaunchDirective = {
            type: 'VideoApp.Launch',
            videoItem: this.options.videoItem,
        };
        return {
            platforms: {
                alexa: {
                    nativeResponse: {
                        response: {
                            directives: [videoAppLaunchDirective],
                            // Responses containing a VideoApp.Launch Directive must not have a shouldEndSession parameter
                            shouldEndSession: undefined,
                        },
                    },
                },
            },
        };
    }
};
VideoAppLaunchOutput = __decorate([
    (0, framework_1.Output)()
], VideoAppLaunchOutput);
exports.VideoAppLaunchOutput = VideoAppLaunchOutput;
