// @jovotech/platform-googleassistant/dist/esm5/output/models/Scene.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { IsEnum, IsNotEmpty, IsObject, IsOptional, IsString, Type, ValidateNested, } from '@jovotech/output';
var NextScene = /** @class */ (function () {
    function NextScene() {
    }
    __decorate([
        IsString(),
        IsNotEmpty(),
        __metadata("design:type", String)
    ], NextScene.prototype, "name", void 0);
    return NextScene;
}());
export { NextScene };
export var SlotFillingStatus;
(function (SlotFillingStatus) {
    SlotFillingStatus["Unspecified"] = "UNSPECIFIED";
    SlotFillingStatus["Initialized"] = "INITIALIZED";
    SlotFillingStatus["Collecting"] = "COLLECTING";
    SlotFillingStatus["Final"] = "FINAL";
})(SlotFillingStatus || (SlotFillingStatus = {}));
var Scene = /** @class */ (function () {
    function Scene() {
    }
    __decorate([
        IsString(),
        __metadata("design:type", String)
    ], Scene.prototype, "name", void 0);
    __decorate([
        IsOptional(),
        IsEnum(SlotFillingStatus),
        __metadata("design:type", String)
    ], Scene.prototype, "slotFillingStatus", void 0);
    __decorate([
        IsObject(),
        __metadata("design:type", Object)
    ], Scene.prototype, "slots", void 0);
    __decorate([
        IsOptional(),
        ValidateNested(),
        Type(function () { return NextScene; }),
        __metadata("design:type", NextScene)
    ], Scene.prototype, "next", void 0);
    return Scene;
}());
export { Scene };
