// @jovotech/platform-googleassistant/dist/esm2015/output/models/content/Card.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { formatValidationErrors, IsEnum, IsNotEmpty, IsOptional, IsSomeValid, isString, IsString, Type, validate, ValidateNested, } from '@jovotech/output';
import { Image, ImageFill } from '../common/Image';
import { Link } from '../common/Link';
export class Card {
    toCard() {
        var _a;
        const card = {
            title: (this.title || this.text || this.subtitle),
        };
        if (this.subtitle) {
            card.subtitle = this.subtitle;
        }
        if (this.text) {
            card.content = this.text;
        }
        if ((_a = this.image) === null || _a === void 0 ? void 0 : _a.url) {
            card.imageUrl = this.image.url;
        }
        return card;
    }
}
__decorate([
    IsOptional(),
    IsString(),
    IsNotEmpty(),
    __metadata("design:type", String)
], Card.prototype, "title", void 0);
__decorate([
    IsOptional(),
    IsString(),
    IsNotEmpty(),
    __metadata("design:type", String)
], Card.prototype, "subtitle", void 0);
__decorate([
    IsSomeValid({
        keys: ['text', 'image'],
        validate: (value) => {
            if (!isString(value)) {
                return '$property must be a string';
            }
            if (!value) {
                return '$property should not be empty';
            }
            return;
        },
    }),
    __metadata("design:type", String)
], Card.prototype, "text", void 0);
__decorate([
    IsSomeValid({
        keys: ['text', 'image'],
        validate: async (value) => {
            if (!(value instanceof Image)) {
                return `$property has to be an instance of Image`;
            }
            const errors = await validate(value);
            if (errors.length) {
                return formatValidationErrors(errors, {
                    text: '$property is invalid:',
                    delimiter: '\n  - ',
                    path: '$property',
                });
            }
            return;
        },
    }),
    Type(() => Image),
    __metadata("design:type", Image)
], Card.prototype, "image", void 0);
__decorate([
    IsOptional(),
    IsEnum(ImageFill),
    __metadata("design:type", String)
], Card.prototype, "imageFill", void 0);
__decorate([
    IsOptional(),
    ValidateNested(),
    Type(() => Link),
    __metadata("design:type", Link)
], Card.prototype, "button", void 0);
