// @jovotech/platform-googleassistant/dist/esm2015/output/decorators/validation/IsValidContentObject.js
import { formatValidationErrors, IsEitherValid, isObject, validate, } from '@jovotech/output';
export function IsValidContentObject(validationOptions) {
    return IsEitherValid({
        name: 'isValidContentObject',
        keys: ['card', 'image', 'table', 'media', 'collection', 'list'],
        validate: async (value, args) => {
            var _a;
            const className = `${args.property.charAt(0).toUpperCase()}${args.property.slice(1)}`;
            if (!isObject(value) || ((_a = value === null || value === void 0 ? void 0 : value.constructor) === null || _a === void 0 ? void 0 : _a.name) !== className) {
                return `$property has to be an instance of ${className}`;
            }
            const errors = await validate(value);
            if (errors.length) {
                return formatValidationErrors(errors, {
                    text: '$property is invalid:',
                    delimiter: '\n  - ',
                    path: '$property',
                });
            }
            return;
        },
    }, validationOptions);
}
